---
title: "Annotation of B lineage lymphocytes"
author: "Jayne Wiarda"
date: "27May2021"
output: 
  github_document: 
    keep_html: TRUE
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.fullwidth=TRUE)
```

We have now processed B lineage lymphocytes from porcine ileum using three methods: cell clustering (and accompanying DGE and hierarchical clustering analyses), topic modeling, and multidimensional DGE analysis. Based on these results, we create final annotations for our cells in this script.

### Load required software packages

The following analyses are conducted using base R packages or the scRNA-seq analysis package, Seurat (Stuart and Butler et al. 2019), ggplot2 (Wickham 2016), scales (Wickham & Seidel 2020), dplyr (Wickham et al. 2020), readxl (Wickham & Bryan 2019). Refer to sessionInfo() at the bottom of the page for the R and package versions used.

```{r, message = FALSE}
library(Seurat)
library(ggplot2)
library(scales)
library(dplyr)
library(readxl)
```

### Import relevant data

Read in our Seurat object from previous analysis:

```{r, warning = FALSE, error = FALSE}
il <- readRDS("/home/Jayne.Wiarda/scRNAseqIleumAtlas/Ileum/Seurat/Ileum_Bonly.rds") # read in corresponding Seurat object
Idents(il) <- il$seurat_clusters
levels(il) <- c('0', '2', '4', '5', '8', '9', '10', '11', '13', '16', '20', '21', '22', '25', '27', '28', '30', '33', '38', '39', '40', '48') 
il$neworder <- Idents(il)
```

## Visualize our cell clusters and their relationships

Start by re-visualizing clusters on a t-SNE plot:

```{r, warning = FALSE, error = FALSE}
DimPlot(il, 
        reduction = 'tsne', 
        group.by = 'neworder',
        label = TRUE)
```

Then re-construct cluster heirarchy:

```{r, warning = FALSE, error = FALSE}
data.tree <- Tool(object = il, 
                  slot = "BuildClusterTree") 
data.tree <- ape::rotateConstr(data.tree, c('33', '25', '9', '13', '30', '8', '5', '40', '39', '22', '16',
                                            '48', '28', '2', '38', '27', '20', '10', '4', '0', '21', '11'))
plot(data.tree, direction = 'downwards', edge.width = 1.5, font = 1)
```

And also re-visualize DGE across clusters:

```{r, warning = FALSE, error = FALSE, message = FALSE}
DE <- read.table('/home/Jayne.Wiarda/scRNAseqIleumAtlas/Ileum/DE/B_OverallDE.txt')
DE <- subset(DE, avg_logFC > 0) # only take genes enriched in the clusters 
Idents(il) <- il$phyloorder
topgenes <- DE %>% group_by(cluster) %>% top_n(5, avg_logFC) # only plot top 5 genes per cluster, as determined by highest average logFC values
DoHeatmap(subset(il, downsample = 100), # take only 100 cells per cluster for plotting
          features = as.character(topgenes$gene), 
          assay = "RNA", 
          disp.min = -1.5, 
          disp.max = 3) +
  scale_fill_gradientn(colors = c("darkturquoise", "grey90", "indianred1", "red"))
```
        
## Visualize our topic model weights & multidimensional DGE gene module scores

Let's read and organize the data:

```{r, warning = FALSE, error = FALSE}
fits <- readRDS('/home/Jayne.Wiarda/scRNAseqIleumAtlas/Ileum/fastTopics/Ileum_Bonly_K_3_K_10_topic_model_fits.rds') # read in all the topic models
cellMem <- as.data.frame(fits$k3$L) # find weighted topic membership for cells when using 3 topics
spatialMod <- as.data.frame(read_excel("/home/Jayne.Wiarda/scRNAseqIleumAtlas/Ileum/Haystack/B_k4_DetectionScores.xlsx", col_types = c(rep('numeric',4), 'text')))
rownames(spatialMod) <- spatialMod$CellBarcodes
spatialMod$CellBarcodes <- NULL
il <- AddMetaData(il, metadata = c(cellMem, spatialMod)) # add topic membership and gene module enrichment to Seurat object
noK <- paste0('k', 1:3)
noMod <- paste0('GeneModule', 1:4)
```

Overlay topic weights onto t-SNE plot:

```{r, warning = FALSE, error = FALSE, message = FALSE}
FeaturePlot(il, features = c(noK), 
            reduction = 'tsne',
            ncol = 3) & 
  scale_color_gradientn( colours = c('grey90', 'darkgreen'),  limits = c(0, 1), oob = squish) & 
  NoAxes() & NoLegend()
```

Overlay gene module detection scores onto t-SNE plot:

```{r, warning = FALSE, error = FALSE, message = FALSE}
FeaturePlot(il, features = c(noMod), 
            reduction = 'tsne',
            ncol = 3) & 
  scale_color_gradientn( colours = c('grey90', 'red4'),  limits = c(0, 0.8), oob = squish) & 
  NoAxes() & NoLegend()
```

## Identify antibody-secreting cells, transitioning B cells, and resting B cells

From looking at cluster-based DGE & hierarchy, we recognize two nodes of cells more distantly related from other clusters. These were clusters 25/33 and clusters 9/13/30. In clusters 25 and 33, we see expression of many genes associated with antibody secretion; however, cluster 33 also has greater expression of some traditional B cell markers and markers of early activation. Moreover, cluster 25 has high weighting for topic 1, enriched with many antibody secretion-associated genes, while cluster 33 has higher weighting for topic 3, associated with B cell activation. Therefore, we termed cluster 25 as antibody-secreting cells (ASCs) and cluster 33 as B cells transitioning into ASCs (transitioning B cells). In clusters 9/13/30, we saw lack of expression for some genes indicitive of BCR rearrangement/cell cycling (e.g. AICDA, PCLAF) but did see expression of other genes that could mark either a memory or naive phenotype (e.g. IFI6, MHC II genes). We also noted these clusters had high expression of genes such as CCR7. These profiles indicated the cells were not currently cycling or undergoing BCR rearrangement, so cells belong to clusters 9/13/30 were termed resting B cells.

Show where ASCs are on t-SNE:

```{r, warning = FALSE, error = FALSE}
ASC <- rownames(il@meta.data %>% filter(il$seurat_clusters == '25'))
DimPlot(il, 
        cells.highlight = ASC, 
        cols.highlight = "red", 
        sizes.highlight = 1,
        cols = "black", 
        order = TRUE, 
        reduction = 'tsne') & NoAxes() & NoLegend() # ASCs in red
```

Show where transitioning B cells are on t-SNE:

```{r, warning = FALSE, error = FALSE}
transB <- rownames(il@meta.data %>% filter(il$seurat_clusters == '33'))
DimPlot(il, 
        cells.highlight = transB, 
        cols.highlight = "red", 
        sizes.highlight = 1,
        cols = "black", 
        order = TRUE, 
        reduction = 'tsne') & NoAxes() & NoLegend() # transitioning B cells in red
```

Show where resting B cells are on t-SNE:

```{r, warning = FALSE, error = FALSE}
restB <- rownames(il@meta.data %>% filter(il$seurat_clusters == '9' | il$seurat_clusters == '13' | il$seurat_clusters == '30'))
DimPlot(il, 
        cells.highlight = restB, 
        cols.highlight = "red", 
        sizes.highlight = 1,
        cols = "black", 
        order = TRUE, 
        reduction = 'tsne') & NoAxes() & NoLegend() # resting B cells in red
```

## Identify cycling B cells

Start by removing our annotated cells from the dataset, so we are now only working with the remaining un-annotated cells:

```{r, warning = FALSE, error = FALSE}
Idents(il) <- il$neworder
il2 <- subset(il, idents = c('33', '25', '30', '13', '9'), invert = TRUE)
```

From looking at genes contributing to topics and gene modules and accompanying GO results, we can see that gene module 3 and topic 2 are enriched for genes typically expressed by cycling cells. Based on this information, let's identify cycling B cells that have high topic 2 weights/gene module 3 detection scores. Remaining cells can be called activated B cells since they express genes associated with activation, such as found in topic 3, but lack expression of cycling genes.

Start by visualizing topic 2 weights vs gene module 3 detection scores, overlaid with topic 3 weights:

```{r, warning = FALSE, error = FALSE}
ggplot(il2@meta.data, aes(x=k2, y=GeneModule3, color = k3)) + 
  geom_point() + 
  scale_colour_gradient(low='grey90', high='darkgreen') +
  theme_classic()
```

Highlight cycling B cells in red as those with topic 2 weights > 0.32 or gene module 3 detection scores > 0.06:

```{r, warning = FALSE, error = FALSE}
plot(il2$k2, il2$GeneModule3, pch = 20, col= ifelse(il2$k2 > 0.32 | il2$GeneModule3 > 0.06, 'red', 'black')) # cycling cells in red
```

Show where cycling cells are found in our t-SNE plot:

```{r, warning = FALSE, error = FALSE}
repB <- rownames(il2@meta.data %>% filter(il2$k2 > 0.32 | il2$GeneModule3 > 0.06))
DimPlot(il2, 
        cells.highlight = repB, 
        cols.highlight = "red", 
        sizes.highlight = 1,
        cols = "black", 
        order = TRUE, 
        reduction = 'tsne') # cycling B cells in red
```

## Identify activated B cells

We identify the remaining cells as activated B cells due to high weighting of topic 3 and expression of genes related to activation.

Show where remaining activated B cells are found in our t-SNE plot:

```{r, warning = FALSE, error = FALSE}
actB <- rownames(il2@meta.data %>% filter(!(il2$k2 > 0.32 | il2$GeneModule3 > 0.06)))
DimPlot(il2, 
        cells.highlight = actB, 
        cols.highlight = "red", 
        sizes.highlight = 1,
        cols = "black", 
        order = TRUE, 
        reduction = 'tsne') # cycling B cells in red
```

## Re-visualize data with new annotations

Now let's add the new annotations to our meta data in the Seurat object:

```{r, warning = FALSE, error = FALSE}
bcs <- as.data.frame(colnames(il))
colnames(bcs) <- 'barcode'
bcs$cellID <- rep('_', nrow(bcs)) # if we left any cells un-annotated, they will show up with an ID of '_'
bcs <- bcs %>% mutate(cellID = replace(cellID, barcode %in% ASC, 'Antibody-secreting cells'))
bcs <- bcs %>% mutate(cellID = replace(cellID, barcode %in% transB, 'Transitioning B cells'))
bcs <- bcs %>% mutate(cellID = replace(cellID, barcode %in% restB, 'Resting B cells'))
bcs <- bcs %>% mutate(cellID = replace(cellID, barcode %in% repB, 'Cycling B cells'))
bcs <- bcs %>% mutate(cellID = replace(cellID, barcode %in% actB, 'Activated B cells'))
write.table(bcs, '/home/Jayne.Wiarda/scRNAseqIleumAtlas/Ileum/Annotations/B.txt')
rownames(bcs) <- bcs$barcode
bcs$barcode <- NULL
il <- AddMetaData(il, metadata = bcs) # add topic membership and gene module enrichment to Seurat object
```

Plot the annotated populations:

```{r, warning = FALSE, error = FALSE}
DimPlot(il, 
        group.by = 'cellID', 
        reduction = 'tsne', 
        cols = c('magenta4', 'darkslateblue', 'mediumpurple1', 'plum4', 'mediumorchid'))
```

And let's also highlight only one cell populations at a time...

ASCs:

```{r, warning = FALSE, error = FALSE}
DimPlot(il, 
        cells.highlight = ASC, 
        cols.highlight = "darkslateblue", 
        sizes.highlight = 1,
        cols = "grey80", 
        order = TRUE, 
        reduction = 'tsne') & NoAxes() & NoLegend()
```

Transitioning B cells:

```{r, warning = FALSE, error = FALSE}
DimPlot(il, 
        cells.highlight = transB, 
        cols.highlight = "mediumorchid", 
        sizes.highlight = 1,
        cols = "grey80", 
        order = TRUE, 
        reduction = 'tsne') & NoAxes() & NoLegend()
```

Activated B cells:

```{r, warning = FALSE, error = FALSE}
DimPlot(il, 
        cells.highlight = actB, 
        cols.highlight = "magenta4", 
        sizes.highlight = 1,
        cols = "grey80", 
        order = TRUE, 
        reduction = 'tsne') & NoAxes() & NoLegend()
```

Cycling B cells:

```{r, warning = FALSE, error = FALSE}
DimPlot(il, 
        cells.highlight = repB, 
        cols.highlight = "mediumpurple1", 
        sizes.highlight = 1,
        cols = "grey80", 
        order = TRUE, 
        reduction = 'tsne') & NoAxes() & NoLegend()
```

Resting B cells:

```{r, warning = FALSE, error = FALSE}
DimPlot(il, 
        cells.highlight = restB, 
        cols.highlight = "plum4", 
        sizes.highlight = 1,
        cols = "grey80", 
        order = TRUE, 
        reduction = 'tsne') & NoAxes() & NoLegend()
```

### Save data

Lastly, let's re-save our Seurat object with the new annotations stored in the cellID meta data slot:

```{r, warning = FALSE, error = FALSE, message = FALSE}
Idents(il) <- il$cellID
saveRDS(il, '/home/Jayne.Wiarda/scRNAseqIleumAtlas/Ileum/Seurat/Ileum_Bonly.rds')
```

### View session information

```{r, warning = FALSE, error = FALSE}
sessionInfo()
```

### References
  
Stuart and Butler et al. Comprehensive Integration of Single-Cell Data. Cell (2019).

H. Wickham. ggplot2: Elegant Graphics for Data Analysis. Springer-Verlag New York, 2016.

Hadley Wickham and Jennifer Bryan (2019). readxl: Read Excel Files. R package version 1.3.1.
  https://CRAN.R-project.org/package=readxl
  
Hadley Wickham and Dana Seidel (2020). scales: Scale Functions for Visualization. R package version 1.1.1.
  https://CRAN.R-project.org/package=scales
  
Hadley Wickham, Romain François, Lionel Henry and Kirill Müller (2020). dplyr: A Grammar of Data Manipulation.
  R package version 1.0.2. https://CRAN.R-project.org/package=dplyr
