---
title: "Annotation of Myeloid Lineage Leukocytes"
author: "Jayne Wiarda"
date: "03June2021"
output: 
  github_document: 
    keep_html: TRUE
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.fullwidth=TRUE)
```

We have now processed myeloid lineage leukocytes from porcine ileum using cell clustering and accompanying DGE and hierarchical clustering analyses. Based on these results, we create final annotations for our cells in this script.

### Load required software packages

The following analyses are conducted using base R packages or the scRNA-seq analysis package, Seurat (Stuart and Butler et al. 2019), ggplot2 (Wickham 2016), scales (Wickham & Seidel 2020), dplyr (Wickham et al. 2020), readxl (Wickham & Bryan 2019). Refer to sessionInfo() at the bottom of the page for the R and package versions used.

```{r, message = FALSE}
library(Seurat)
library(ggplot2)
library(scales)
library(dplyr)
library(readxl)
```

### Import relevant data

Read in our Seurat object from previous analysis:

```{r, warning = FALSE, error = FALSE}
il <- readRDS('/home/Jayne.Wiarda/scRNAseqIleumAtlas/Ileum/Seurat/Ileum_MyeloidOnly.rds')
Idents(il) <- il$seurat_clusters
```

## Visualize our cell clusters and their relationships

Start by re-visualizing clusters on a t-SNE plot:

```{r, warning = FALSE, error = FALSE}
DimPlot(il, 
        reduction = 'tsne', 
        group.by = 'neworder',
        label = TRUE)
```

Then re-construct cluster heirarchy:

```{r, warning = FALSE, error = FALSE}
data.tree <- Tool(object = il, 
                  slot = "BuildClusterTree") 
data.tree <- ape::rotateConstr(data.tree, c('42', '49', '52'))
plot(data.tree, direction = 'downwards', edge.width = 1.5, font = 1)
```

And also re-visualize DGE across clusters:

```{r, warning = FALSE, error = FALSE, message = FALSE}
DE <- read.table('/home/Jayne.Wiarda/scRNAseqIleumAtlas/Ileum/DE/MyeloidOnly_OverallDE.txt')
DE <- subset(DE, avg_logFC > 0) # only take genes enriched in the clusters 
Idents(il) <- il$phyloorder
topgenes <- DE %>% group_by(cluster) %>% top_n(5, avg_logFC) # only plot top 5 genes per cluster, as determined by highest average logFC values
DoHeatmap(subset(il, downsample = 100), # take only 100 cells per cluster for plotting
          features = as.character(topgenes$gene), 
          assay = "RNA", 
          disp.min = -1.5, 
          disp.max = 3) +
  scale_fill_gradientn(colors = c("darkturquoise", "grey90", "indianred1", "red"))
```
        
## Visualize other key canonical genes

Let's also visualize expression of a curated list of canonical genes for different myeloid lineage leukocytes.

Dendritic cell genes: SLA-DQB1, SLA-DRA, CD74, FLT3
Macrophage genes: CD68, CD14, TLR4, NLRP3, CSF1R, CD163
Mast cell genes: MS4A2, FCER1A, IL13

Visualize in dot plot:

```{r, warning = FALSE, error = FALSE, message = FALSE}
DotPlot(il, 
        features = c('SLA-DQB1', 'HLA-DRA', 'CD74', 'FLT3', # high in DCs
                     'CD68', 'CD14', 'TLR4', 'NLRP3', 'CSF1R', 'CD163', # high in macrophages
                     'MS4A2', 'FCER1A', 'IL13'), # high in mast cells
        cols = c('gold', 'red')) + RotatedAxis()
```

Visualize in feature plots:

```{r, warning = FALSE, error = FALSE, message = FALSE}
FeaturePlot(il,
            features = c('SLA-DQB1', 'HLA-DRA', 'CD74', 'FLT3', # high in DCs
                     'CD68', 'CD14', 'TLR4', 'NLRP3', 'CSF1R', 'CD163', # high in macrophages
                     'MS4A2', 'FCER1A', 'IL13'), # high in mast cells
            reduction = 'tsne',
            cols = c("grey90", "purple4"),
            ncol = 3) & NoLegend() & NoAxes()
```

These gene expression profiles indicate cluster 42 to be primarily DCs, cluster 49 macrophages, and cluster 52 mast cells.

## Re-visualize data with new annotations

Now let's add the new annotations to our meta data in the Seurat object:

```{r, warning = FALSE, error = FALSE}
bcs <- as.data.frame(colnames(il))
colnames(bcs) <- 'barcode'
bcs$cellID <- rep('_', nrow(bcs))
df <- as.data.frame(il$seurat_clusters)
# Cluster-based cell annotations not provided in the annotation .txt files:
DC <- rownames(subset(df, il$seurat_clusters == '42'))
mac <- rownames(subset(df, il$seurat_clusters == '49'))
mast <- rownames(subset(df, il$seurat_clusters == '52'))
bcs <- bcs %>% mutate(cellID = replace(cellID, barcode %in% DC, 'Dendritic cells'))
bcs <- bcs %>% mutate(cellID = replace(cellID, barcode %in% mac, 'Macrophages'))
bcs <- bcs %>% mutate(cellID = replace(cellID, barcode %in% mast, 'Mast cells'))
rownames(bcs) <- bcs$barcode
bcs$barcode <- NULL
il <- AddMetaData(il, metadata = bcs) # add new annotations to meta data slot
```

Let's plot new annotations in a t-SNE plot:

```{r, warning = FALSE, error = FALSE}
DimPlot(il, 
        group.by = 'cellID', 
        reduction = 'tsne', 
        cols = c('cornflowerblue', 'cyan4', 'blue'))
```

And let's also highlight only one cell populations at a time...

DCs:

```{r, warning = FALSE, error = FALSE}
DimPlot(il, 
        cells.highlight = DC, 
        cols.highlight = "cornflowerblue", 
        sizes.highlight = 1,
        cols = "grey80", 
        order = TRUE, 
        reduction = 'tsne') & NoAxes() & NoLegend()
```

Macrophages:

```{r, warning = FALSE, error = FALSE}
DimPlot(il, 
        cells.highlight = mac, 
        cols.highlight = "cyan4", 
        sizes.highlight = 1,
        cols = "grey80", 
        order = TRUE, 
        reduction = 'tsne') & NoAxes() & NoLegend()
```

Mast cells:

```{r, warning = FALSE, error = FALSE}
DimPlot(il, 
        cells.highlight = mast, 
        cols.highlight = "blue", 
        sizes.highlight = 1,
        cols = "grey80", 
        order = TRUE, 
        reduction = 'tsne') & NoAxes() & NoLegend()
```

### Save data

Lastly, let's re-save our Seurat object with the new annotations stored in the cellID meta data slot:

```{r, warning = FALSE, error = FALSE, message = FALSE}
Idents(il) <- il$cellID
saveRDS(il, '/home/Jayne.Wiarda/scRNAseqIleumAtlas/Ileum/Seurat/Ileum_MyeloidOnly.rds')
```

### View session information

```{r, warning = FALSE, error = FALSE}
sessionInfo()
```

### References

Stuart and Butler et al. Comprehensive Integration of Single-Cell Data. Cell (2019).

H. Wickham. ggplot2: Elegant Graphics for Data Analysis. Springer-Verlag New York, 2016.

Hadley Wickham and Jennifer Bryan (2019). readxl: Read Excel Files. R package version 1.3.1.
  https://CRAN.R-project.org/package=readxl
  
Hadley Wickham and Dana Seidel (2020). scales: Scale Functions for Visualization. R package version 1.1.1.
  https://CRAN.R-project.org/package=scales
  
Hadley Wickham, Romain François, Lionel Henry and Kirill Müller (2020). dplyr: A Grammar of Data Manipulation.
  R package version 1.0.2. https://CRAN.R-project.org/package=dplyr
