---
title: "Characterization of Ileal Cells: General Characterization"
author: "Jayne Wiarda"
date: "31Mar2021"
output: 
  github_document: 
    keep_html: TRUE
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.fullwidth=TRUE)
```

We have now processed scRNA-seq data from six porcine ileum samples through general data QC and pre-processing steps. At this point, we can start digging into the data more in order to form some biological interpretations! By the end of this analysis, we will be able to classify our cells into some general lineage designations, and we will also be able to see some biological differences occurring between different sample types. We will also be able to output lists of genes differentially expressed amongst clusters.

Sample nomenclature is as follows:

* IPP = ileum with only Peyer's patch regions present
* NoPP = ileum with only regions without Peyer's patches present
* X2I = ileum with both regions with and without Peyer's patches present; corresponding to a whole ileum cross section
* Pig 1 = IPP1, NoPP1, X2I1
* Pig 2 = IPP2, NoPP2, X2I2

Analyses are divided into sections as follows, and only one of these components is specified and included below:

* General characterization of the data, where we visualize various metrics for our cells and/or clusters
* Hierarchical clustering of the data to determine phylogenetic relationship amongst cell clusters
* Cluster-based differential gene expression analysis to determine genes differentially expressed in a cluster relative to the rest of the dataset (cluster-based overall DGE) or in one cluster relative to another (cluster-based pairwise DGE)
* Cluster-based pseudobulk correlation as another method to assess similarity/dissimilarity between pairwise cluster combinations
* Tissue composition comparisons, to determine differences in the compositions of different cells and/or transcripts across different ileal sample types
* Sample-specific quality comparisons, where we assess the whole transcriptomic profiles from each sample
* Cluster-based gene set enrichment analysis, where we calculate enrichment of signatures for sorted immune cell populations from bulk RNA-seq datasets within our single-cell dataset

This file specifically goes over general characterization of our dataset.

## General characterization of the data

### Load required software packages

The following analyses are conducted using base R packages or the scRNA-seq analysis package, Seurat (Stuart and Butler et al. 2019). Refer to sessionInfo() at the bottom of the page for the R and package versions used.

```{r, message = FALSE}
library(Seurat)         
```

### Import relevant data 

Read in the RDS file from previous work:

```{r, warning = FALSE, error = FALSE}
il <- readRDS("/home/Jayne.Wiarda/scRNAseqIleumAtlas/Ileum/Seurat/IleumAtlasAll.rds") 
DefaultAssay(il) <- "RNA" # set our default assay to RNA for subsequent work
Idents(il) <- il$seurat_clusters # change our default IDs to our cluster assignments
```

Also bring in the number of 'significant' PCs that we calculated for this dataset previously:

```{r, warning = FALSE, error = FALSE}
PCdims <- 1:19
```

### Visualize the data

Visualize by UMAP dimensionality reduction:

```{r, warning = FALSE, error = FALSE}
DimPlot(il, 
        label = TRUE,
        reduction = 'umap',
        group.by = "seurat_clusters",
        label.size = 4) 
```

Visualize by t-SNE dimensionality reduction:
```{r, warning = FALSE, error = FALSE}
DimPlot(il, 
        label = TRUE,
        reduction = 'tsne',
        group.by = "seurat_clusters",
        label.size = 4) 
```
In our future analyses, we will utilize the UMAP plot to visualize our entire ileum dataset, as UMAP tends to capture global relationships better than t-SNE.

### Further QC visualization

Let's look at some of our parameters that were used during our cell filtering QC steps. We can visualize these overlaid onto our UMAP and as violin plots.

Doublet probability scores:

```{r, warning = FALSE, error = FALSE}
FeaturePlot(il, features = c("Scrublet"), pt.size = 0.5, min.cutoff = .05, max.cutoff = 0.25) # view doublet probabilities
```

```{r, warning = FALSE, error = FALSE, fig.height = 6, fig.width = 15}
VlnPlot(il, features = c("Scrublet"), pt.size = 0.5)
```

Percent mitochondrial reads:

```{r, warning = FALSE, error = FALSE}
FeaturePlot(il, features = c("prcntMito"), pt.size = 0.5, min.cutoff = .025, max.cutoff = 0.1) # view % mitochondrial reads
```

```{r, warning = FALSE, error = FALSE, fig.height = 6, fig.width = 15}
VlnPlot(il, features = c("prcntMito"), pt.size = 0.5)
```

Total UMIs per cell:

```{r, warning = FALSE, error = FALSE}
FeaturePlot(il, features = c("UmiSums"), pt.size = 0.5, min.cutoff = 5000, max.cutoff = 20000) # view total UMIs
```

```{r, warning = FALSE, error = FALSE, fig.height = 6, fig.width = 15}
VlnPlot(il, features = c("UmiSums"), pt.size = 0.5)
```

Total transcripts per cell:

```{r, warning = FALSE, error = FALSE}
FeaturePlot(il, features = c("nCount_RNA"), pt.size = 0.5, min.cutoff = 5000, max.cutoff = 30000) # view total RNA counts
```

```{r, warning = FALSE, error = FALSE, fig.height = 6, fig.width = 15}
VlnPlot(il, features = c("nCount_RNA"), pt.size = 0.5)
```

Total genes detected per cell:

```{r, warning = FALSE, error = FALSE}
FeaturePlot(il, features = c("GenesDetected"), pt.size = 0.5, min.cutoff = 500, max.cutoff = 3000) # view total genes detected
```

```{r, warning = FALSE, error = FALSE, fig.height = 6, fig.width = 15}
VlnPlot(il, features = c("GenesDetected"), pt.size = 0.5)
```

### Assess canonical gene expression

Let's see if we can broadly identify our cells by assessing expression patterns for some canonical genes we know from previous literature.

Start by overlaying expression profiles onto our UMAP plot:

```{r, warning = FALSE, error = FALSE, fig.height = 20, fig.width = 15}
FeaturePlot(il,
            features = c('CD19', 'CD79B', 'MS4A1', 'JCHAIN', # B/plasma cell genes
                         'CD3E', 'CD3G', 'CD52', 'ZAP70', # T/ILC genes
                         'ENSSSCG00000028461', 'CD68', 'CSF2RB', 'ICAM1',  # myeloid lineage leukocyte genes
                         'PTPRC', # pan-immune cell gene
                         'EPCAM', 'KRT8', # epithelial cell genes
                         'ACTA2', 'COL18A1'), # endothelial cell genes 
            cols = c("grey90", "purple4")) & NoLegend() & NoAxes()
```

To determine how genes are expressed within our Seurat clusters, let's create a dot plot to summarize how many cells express a gene in each cluster and at what relative level:

```{r, warning = FALSE, error = FALSE, fig.height = 12}
DotPlot(il, 
        features = c('CD19', 'CD79B', 'MS4A1', 'JCHAIN', # B/plasma cell genes
                     'CD3E', 'CD3G', 'CD52', 'ZAP70', # T/ILC genes
                     'ENSSSCG00000028461', 'CD68', 'CSF2RB', 'ICAM1', # non-lymphoid immune cell genes
                     'EPCAM', 'KRT8', 'ACTA2', 'COL18A1',
                     'PTPRC'), # non-immune cell genes
        cols = c("gold", "red"), 
        col.min = -1, 
        col.max = 2,
        dot.scale = 6,
        assay = "RNA",
        group.by = "seurat_clusters") +
  RotatedAxis()
```

Based off assessing expression patterns for these genes, we can broadly classify our cells into cell lineages as follows:

* B lineage lymphocytes (clusters 0, 2, 4, 5, 8, 9, 10, 11, 13, 16, 20, 21, 22, 25, 27, 28, 30, 33, 38, 39, 40, 48)
* T/ILC lineage lymphocytes (clusters 1, 3, 6, 7, 12, 14, 15, 17, 18, 19, 23, 24, 26, 29, 31, 32, 34, 35, 37, 41, 43, 44, 46, 47, 51, 53)
* Myeloid lineage leukocytes (clusters 42, 49, 52)
* Non-leukocytes (clusters 36, 45, 50)

With these putative cell lineage identities in mind, we may want to organize our data into these groupings for later analyses. To make this easier later, let's store that information in a metadata slot for our Seurat object, called 'neworder':

```{r, warning = FALSE, error = FALSE, fig.height = 12}
Idents(il) <- il$seurat_clusters
levels(il) <- c('0', '2', '4', '5', '8', '9', '10', '11', '13', '16', '20', '21',
                '22', '25', '27', '28', '30', '33', '38', '39', '40', '48',
                '1', '3', '6', '7', '12', '14', '15', '17', '18', '19', '23', 
                '24', '26', '29', '31', '32', '34', '35', '37', '41', '43', 
                '44', '46', '47', '51', '53', 
                '42', '49', '52', 
                '36', '45', '50') 
il$neworder <- Idents(il)
Idents(il) <- il$neworder
```

Based on the new order, we can again replot our data:

```{r, warning = FALSE, error = FALSE, fig.height = 12}
DotPlot(il, 
        features = c('CD19', 'CD79B', 'MS4A1', 'JCHAIN', # B/plasma cell genes
                     'CD3E', 'CD3G', 'CD52', 'ZAP70', # T/ILC genes
                     'ENSSSCG00000028461', 'CD68', 'CSF2RB', 'ICAM1', # non-lymphoid immune cell genes
                     'EPCAM', 'KRT8', 'ACTA2', 'COL18A1',
                     'PTPRC'), # non-immune cell genes
        cols = c("gold", "red"), 
        col.min = -1, 
        col.max = 2,
        dot.scale = 6,
        assay = "RNA",
        group.by = "neworder") +
  RotatedAxis()
```

We can also implement our new lineage assignments to create a new meta data slot called 'celltypes':

```{r, warning = FALSE, error = FALSE}
Idents(il) <- il$neworder
B <- rep('B lymphoid lineage leukocytes', 22)
TILC <- rep('T/ILC lymphoid lineage leukocytes', 26)
Myeloid <- rep('Myeloid lineage leukocytes', 3)
notImmune <- rep('Non-leukocytes', 3)
GutCellTypes <- c(B, TILC, Myeloid, notImmune)
length(GutCellTypes)
il$celltypes <-il$neworder
Idents(il) <- il$celltypes
names(GutCellTypes) <- levels(il) # assign GutCellTypes to cluster numbers
il <- RenameIdents(il, GutCellTypes) # change dataset identity to cell types in new Seurat object
il$celltypes <- Idents(il)
```

And we can further visualize our cell type designations on plots.

UMAP:

```{r, warning = FALSE, error = FALSE}
DimPlot(il, 
        reduction = "umap", 
        cols = c("mediumorchid", "orange", "blue", "green3"), 
        pt.size = .25)
```

t-SNE:

```{r, warning = FALSE, error = FALSE}
DimPlot(il, 
        reduction = "tsne", 
        cols = c("mediumorchid", "orange", "blue", "green3"), 
        pt.size = .25)
```
        
### Assess cluster/cell distributions

Based on the clusters we've identified, let's next find out how these are distributed across our dataset. 

First, let's determine some basic information about our clusters, such as how many cells are in each cluster:

```{r, warning = FALSE, error = FALSE}
cluster_no <- table(il$seurat_clusters) # how many cells are in each cluster?
cluster_no
sum(cluster_no) # how many cells are there total?
min(cluster_no) # how many cells are in our smallest cluster?
```

Let's go on to plot these cluster sizes, taking into account also which sample cells were derived from:

```{r, warning = FALSE, error = FALSE, fig.width = 14}
clusterSample_no <- table(il$SampleID, il$seurat_clusters)
barplot(clusterSample_no,
        col = c("chartreuse4", "chartreuse3", "darkorange3", 
                "darkorange", "deepskyblue4", "deepskyblue3"),,
        xlab = "Cluster #", 
        ylab = "Number of cells", 
        las = 2,
        legend = rownames(clusterSample_no),
        border = NA,
        space = 0.05)
```

Instead of looking at the size of clusters, let's look at the proportions of cells coming from different samples within each cluster. 

Look at the proportion of cells coming from each sample type (IPP, NoPP, X2I) within each cluster:

```{r, warning = FALSE, error = FALSE, fig.width = 14}
Idents(il) <- il$neworder
il$tissue <- substring(il$SampleID, 1, 3) # create a tissue assignment for each cell
GutTissueTotalCells <- prop.table(table(il$tissue)) # What percent of total cells are from each tissue?
GutTissuePercents <- prop.table(table(Idents(il),il$tissue), 
                                margin = 1) # What percent of cells from each cluster belong to each tissue?
GutTissuePercents <- rbind(GutTissuePercents, GutTissueTotalCells) # add row of overall percentages to table
#rowSums(GutTissuePercents) # make sure all row sums are equal to 1
GutTissuePercents <- t(GutTissuePercents) # transpose the table
par(mfrow=c(1, 1), mar=c(5, 5, 4, 8))
barplot(GutTissuePercents, # create stacked bar plot
        col = c("chartreuse3", "darkorange", "deepskyblue3"), 
        legend = rownames(GutTissuePercents),
        xlab = "Cluster #", 
        ylab = "Frequency within cluster", 
        las = 2,
        border = NA,
        space = 0.05,
        legend.text = TRUE, 
        args.legend = list(x = "topright", bty = "n", inset=c(-0.15, 0)))
```

Look at the proportion of cells coming from each animal (Pig 1, Pig 2) within each cluster:

```{r, warning = FALSE, error = FALSE, fig.width = 14}
Idents(il) <- il$neworder
il$pigID <- substr(il$SampleID, 
                   nchar(il$SampleID) - 1 + 1, 
                   nchar(il$SampleID)) # create pig number assignment for each cell based on last character in SampleID assignment
GutPigTotalCells <- prop.table(table(il$pigID)) # What percent of total cells are from each pig?
GutPigPercents <- prop.table(table(Idents(il),il$pigID), 
                             margin = 1) # What percent of cells from each cluster belong to each tissue?
GutPigPercents <- rbind(GutPigPercents, GutPigTotalCells) # add row of overall percentages to table
#rowSums(GutPigPercents) # make sure all row sums are equal to 1
GutPigPercents <- t(GutPigPercents) # transpose the table
par(mfrow=c(1, 1), mar=c(5, 5, 4, 8))
barplot(GutPigPercents, # create stacked bar plot
        col = c("red", "dodgerblue3"), 
        legend = rownames(GutPigPercents),
        xlab = "Cluster #", 
        ylab = "Frequency within cluster", 
        las = 2,
        border = NA,
        space = 0.05,
        legend.text = TRUE, 
        args.legend = list(x = "topright", bty = "n", inset=c(-0.15, 0)))
```

Look at the proportion of cells coming from each individual sample (IPP1, IPP2, NoPP1, NoPP2, X2I1, X2I2) within each cluster:

```{r, warning = FALSE, error = FALSE, fig.width = 14}
Idents(il) <- il$neworder
GutSampleTotalCells <- prop.table(table(il$SampleID)) # What percent of total cells are from each sample?
GutSamplePercents <- prop.table(table(Idents(il),il$SampleID), 
                                margin = 1) # What percent of cells from each cluster belong to each sample?
GutSamplePercents <- rbind(GutSamplePercents, GutSampleTotalCells) # add row of overall percentages to table
#rowSums(GutSamplePercents) # make sure all are equal to 1
GutSamplePercents <- t(GutSamplePercents) # transpose the table
par(mfrow=c(1, 1), mar=c(5, 5, 4, 8))
barplot(GutSamplePercents, # create stacked bar plot
        col = c("chartreuse4", "chartreuse3", "darkorange3", 
                "darkorange", "deepskyblue4", "deepskyblue3"), 
        legend = rownames(GutSamplePercents),
        xlab = "Cluster #", 
        ylab = "Frequency within cluster", 
        las = 2,
        border = NA,
        space = 0.05,
        legend.text = TRUE, 
        args.legend = list(x = "topright", bty = "n", inset=c(-0.15, 0)))
```

We can also view the three previous metrics overlaid onto our UMAP plot.

Tissue ID:

```{r, warning = FALSE, error = FALSE}
Idents(il) <- il$seurat_clusters
DimPlot(il, # make UMAP showing sample origins
        reduction = "umap", 
        pt.size = 0.1, 
        group.by = "tissue")
```

```{r, warning = FALSE, error = FALSE, fig.width = 20}
DimPlot(il, # make UMAP showing sample origins
        reduction = "umap", 
        pt.size = 0.1, 
        split.by = "tissue")
```

Animal ID:

```{r, warning = FALSE, error = FALSE}
Idents(il) <- il$seurat_clusters
DimPlot(il, # make UMAP showing sample origins
        reduction = "umap", 
        pt.size = 0.1, 
        group.by = "pigID")
```

```{r, warning = FALSE, error = FALSE, fig.width = 15}
DimPlot(il, # make UMAP showing sample origins
        reduction = "umap", 
        pt.size = 0.1, 
        split.by = "pigID")
```

Sample IDs:

```{r, warning = FALSE, error = FALSE, fig2, fig.height = 3, fig.width = 6}
Idents(il) <- il$seurat_clusters
DimPlot(il, # make UMAP showing sample origins
        reduction = "umap", 
        pt.size = 0.1, 
        group = "SampleID")
```

```{r, warning = FALSE, error = FALSE, fig.width = 28}
DimPlot(il, # make UMAP showing sample origins
        reduction = "umap", 
        pt.size = 0.1, 
        split.by = "SampleID")
```

### Save our data

Since we've added some new meta data information to our dataset, let's save our Seurat object as an .rds again:

```{r, warning = FALSE, error = FALSE, eval = TRUE}
saveRDS(il, "/home/Jayne.Wiarda/scRNAseqIleumAtlas/Ileum/Seurat/IleumAtlasAll.rds")
```

### View session information

```{r, warning = FALSE, error = FALSE}
sessionInfo()
```

### References

Stuart and Butler et al. Comprehensive Integration of Single-Cell Data. Cell (2019).
