---
title: "Characterization of Ileal Cells: Pairwise Differential Gene Expression (DGE) Analysis"
author: "Jayne Wiarda"
date: "31Mar2021"
output: 
  github_document: 
    keep_html: TRUE
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.fullwidth=TRUE)
```

We have now processed scRNA-seq data from six porcine ileum samples through general data QC and pre-processing steps. At this point, we can start digging into the data more in order to form some biological interpretations! By the end of this analysis, we will be able to classify our cells into some general lineage designations, and we will also be able to see some biological differences occurring between different sample types. We will also be able to output lists of genes differentially expressed amongst clusters.

Sample nomenclature is as follows:

* IPP = ileum with only Peyer's patch regions present
* NoPP = ileum with only regions without Peyer's patches present
* X2I = ileum with both regions with and without Peyer's patches present; corresponding to a whole ileum cross section
* Pig 1 = IPP1, NoPP1, X2I1
* Pig 2 = IPP2, NoPP2, X2I2

Analyses are divided into sections as follows, and only one of these components is specified and included below:

* General characterization of the data, where we visualize various metrics for our cells and/or clusters
* Hierarchical clustering of the data to determine phylogenetic relationship amongst cell clusters
* Cluster-based differential gene expression analysis to determine genes differentially expressed in a cluster relative to the rest of the dataset (cluster-based overall DGE) or in one cluster relative to another (cluster-based pairwise DGE)
* Cluster-based pseudobulk correlation as another method to assess similarity/dissimilarity between pairwise cluster combinations
* Tissue composition comparisons, to determine differences in the compositions of different cells and/or transcripts across different ileal sample types
* Sample-specific quality comparisons, where we assess the whole transcriptomic profiles from each sample
* Cluster-based gene set enrichment analysis, where we calculate enrichment of signatures for sorted immune cell populations from bulk RNA-seq datasets within our single-cell dataset

This file specifically goes over cluster-based differential gene expression analysis to determine genes differentially expressed in one cluster relative to another (cluster-based pairwise DGE).

## Pairwise DGE analysis of cell clusters

We perform pairwise DGE analysis for every pairwise combination of clusters.

Our criteria for a differentially expressed gene are as follows:

* logFC value > 0.25 
* gene expressed in at least 10% of cells within one of the two populations being compared
  - Population 1 = cluster A of pairwise comparison
  - Population 2 = cluster B of pairwise comparison
* adjusted p-value < 0.05

We use the default (Wilcoxon) method to determine DGE and report logFC as the natural log.

### Load required software packages

The following analyses are conducted using base R packages or the scRNA-seq analysis package, Seurat (Stuart and Butler et al. 2019). To run some of these functions more quickly, we also utilize parallelization with the software package, future (Bengtsson 2020). Refer to sessionInfo() at the bottom of the page for the R and package versions used.

Other required packages: ggplot2 (Wickham 2016), writexl (Ooms 2020), scales (Wickham & Seidel 2020)

```{r, message = FALSE}
library(Seurat)  
library(writexl)        
library(scales)
library(ggplot2)
library(future)
```

### Import relevant data 

Read in the RDS file from previous work:

```{r, warning = FALSE, error = FALSE}
il <- readRDS("/home/Jayne.Wiarda/scRNAseqIleumAtlas/Ileum/Seurat/IleumAtlasAll.rds") 
DefaultAssay(il) <- "RNA"
Idents(il) <- il$phyloorder # place clusters in phylo order determined by hierarchical clustering analysis
```

### Allow parallelization 

```{r, warning = FALSE, error = FALSE}
plan("multiprocess", workers = 7) # change plan to allow parallelization with 7 workers
```

### Determine pairwise comparisons

First, we need to set up all of our pairwise combinations, stored as 'comps':

```{r, warning = FALSE, error = FALSE, message = FALSE}
clusters <- unique(Idents(il)) # identify all of our cluster IDs
pairwise <- combn(clusters, 2) # create all pairwise combinations of cluster IDs
p1 <- pairwise[1,] 
p2 <- pairwise[2,] 
comps1 <- data.frame(p1, p2)
colnames(comps1) <- c('pop1', 'pop2')
comps2 <- data.frame(p2, p1)
colnames(comps2) <- c('pop1', 'pop2')
comps <- rbind(comps1, comps2)
```

### Perform pairwise DGE analyses

```{r, warning = FALSE, error = FALSE, message =FALSE}
results <- list()
for(i in 1:nrow(comps)) {
  markers <- FindMarkers(il, 
                         ident.1 = comps[i,1], 
                         ident.2 = comps[i,2],
                         assay = "RNA",
                         only.pos = TRUE,
                         logfc.threshold = 0.25, # minimum logFC of 0.25
                         min.pct = 0.1) # expressed in at least 10% of all cells in the cluster or entire dataset
  markers$gene <- rownames(markers)
  markers$pop1 <- paste(comps[i,1])
  markers$pop2 <- paste(comps[i,2])
  markers$comparison <- paste(markers$pop1, markers$pop2, sep = 'v')
  results[[i]] <- markers
} # if any of the comparisons don't turn up DE genes, this function won't work... it's then also likely the data has been over-clustered in preceding steps....
pwAll <- do.call(rbind, results)
pwAll <- subset(pwAll, p_val_adj < 0.05)
```

We also add in additional gene symbol and Ensembl ID information:

```{r, warning = FALSE, error = FALSE, message =FALSE}
features <- read.table("/home/Jayne.Wiarda/scRNAseqIleumAtlas/Ileum/QC/GeneInfo.txt", 
                       header = TRUE) # read in the gene information for the filtered gene list from QC analysis
head(features, n = 3) # check the column names
features <- features[c("Symbol", "EnsemblID", "Name")] # subset only the columns of gene symbols, Ensembl IDs, and the names used for analysis
head(features, n = 3) # check that subset worked
pwAll <- merge(pwAll, # merge the DE gene lists with the additional gene information
               features, 
               by.x = "gene", 
               by.y = "Name")
pwAll <- pwAll[order(pwAll$comp, pwAll$p_val_adj),] # reorder by lowest to highest p-value within each cluster
```

### Save our differentially expressed gene lists

```{r, warning = FALSE, error = FALSE}
write.table(pwAll, 
            file = "/home/Jayne.Wiarda/scRNAseqIleumAtlas/Ileum/DE/AllClusters_PairwiseDGE.txt") 
#write_xlsx(x = pwAll, 
#          path = "/home/Jayne.Wiarda/scRNAseqIleumAtlas/Ileum/DE/AllClusters_PairwiseDGE.xlsx",
#          col_names = TRUE) # this command won't work if more than ~1.048 million rows in pwAll exist; which is the case here
```

### Make a heatmap of the top DE genes

Calculate the number of DE genes with significantly increased versus decreased expression for each pairwise cluster comparison:

```{r, warning = FALSE, error = FALSE}
results <- list()
for(i in 1:nrow(comps)) {
  DE <- sum(pwAll$pop1 == comps[i,1] & pwAll$pop2 == comps[i,2])
  DE$pop1 <- paste(comps[i,1])
  DE$pop2 <- paste(comps[i,2])
  results[[i]] <- DE
}
pwDE <- data.frame(matrix(unlist(results), nrow=length(results), byrow=T))
colnames(pwDE) <- c('DEgenes', 'pop1', 'pop2')
pwDE$DEgenes <- as.numeric(pwDE$DEgenes)
pwDE$pop1 <- factor(pwDE$pop1,levels = c(levels(il$phyloorder)))
pwDE$pop2 <- factor(pwDE$pop2,levels = c(levels(il$phyloorder)))
```

Save the information pertaining to the number of DE genes with significantly increased versus decreased expression in each calculation:

```{r, warning = FALSE, error = FALSE}
write_xlsx(x = pwDE, 
           path = "/home/Jayne.Wiarda/scRNAseqIleumAtlas/Ileum/DE/AllClusters_PairwiseDGE_GeneSumsUpDown.xlsx",
           col_names = TRUE)
```

Create a heatmap of genes with significantly increased versus decreased expression in different pairwise cluster comparisons:

```{r, warning = FALSE, error = FALSE, fig.width = 35, fig.height = 25}
ggplot(pwDE, aes(pop1, pop2, fill = DEgenes)) +
  geom_tile()+
  scale_fill_gradientn(colours = c('beige', 'yellow','orange','red', 'darkred'),
                       limits = c(0, 1000), oob=squish)+ 
  theme_classic()
```

Calculate the total number of DE genes for each pairwise cluster comparison:

```{r, warning = FALSE, error = FALSE}
results <- list()
for(i in 1:nrow(comps)) {
  DE <- sum(pwAll$pop1 == comps[i,1] & pwAll$pop2 == comps[i,2] | pwAll$pop1 == comps[i,2] & pwAll$pop2 == comps[i,1])
  DE$pop1 <- paste(comps[i,1])
  DE$pop2 <- paste(comps[i,2])
  results[[i]] <- DE
}
pwDEcombined <- data.frame(matrix(unlist(results), nrow=length(results), byrow=T))
colnames(pwDEcombined) <- c('DEgenes', 'pop1', 'pop2')
pwDEcombined$DEgenes <- as.numeric(pwDEcombined$DEgenes)
pwDEcombined$pop1 <- factor(pwDEcombined$pop1,levels = c(levels(il$phyloorder)))
pwDEcombined$pop2 <- factor(pwDEcombined$pop2,levels = c(levels(il$phyloorder)))
```

Save the information pertaining to the total number of DE genes in each calculation:

```{r, warning = FALSE, error = FALSE}
write_xlsx(x = pwDEcombined, 
           path = "/home/Jayne.Wiarda/scRNAseqIleumAtlas/Ileum/DE/AllClusters_PairwiseDGE_GeneSumsCombined.xlsx",
           col_names = TRUE)
```

Create a heatmap of genes with total DE genes in different pairwise cluster comparisons:

```{r, warning = FALSE, error = FALSE, fig.width = 35, fig.height = 25}
ggplot(pwDEcombined, aes(pop1, pop2, fill = DEgenes)) +
  geom_tile()+
  #scale_fill_gradientn(colours = c('slateblue4', 'violetred', 'orange', 'gold'))+ 
  theme_classic()+
  scale_fill_gradientn(colours = c('beige', 'yellow','orange','red', 'darkred'),
                       limits = c(0, 1500), oob=squish)
```

### View session information

```{r, warning = FALSE, error = FALSE}
sessionInfo()
```

### References

H. Bengtsson, A Unifying Framework for Parallel and Distributed Processing in R using Futures, arXiv:2008.00553, 2020

Jeroen Ooms (2020). writexl: Export Data Frames to Excel 'xlsx' Format. R package version 1.3.1. https://CRAN.R-project.org/package=writexl
  
Stuart and Butler et al. Comprehensive Integration of Single-Cell Data. Cell (2019).

H. Wickham. ggplot2: Elegant Graphics for Data Analysis. Springer-Verlag New York, 2016.
 
Hadley Wickham and Dana Seidel (2020). scales: Scale Functions for Visualization. R package version 1.1.1. https://CRAN.R-project.org/package=scales
