---
title: "Characterization of Ileal Cells: Cell Lineage Comparisons Across Tissues"
author: "Jayne Wiarda"
date: "31Mar2021"
output: 
  github_document: 
    keep_html: TRUE
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.fullwidth=TRUE)
```

We have now processed scRNA-seq data from six porcine ileum samples through general data QC and pre-processing steps. At this point, we can start digging into the data more in order to form some biological interpretations! By the end of this analysis, we will be able to classify our cells into some general lineage designations, and we will also be able to see some biological differences occurring between different sample types. We will also be able to output lists of genes differentially expressed amongst clusters.

Sample nomenclature is as follows:

* IPP = ileum with only Peyer's patch regions present
* NoPP = ileum with only regions without Peyer's patches present
* X2I = ileum with both regions with and without Peyer's patches present; corresponding to a whole ileum cross section
* Pig 1 = IPP1, NoPP1, X2I1
* Pig 2 = IPP2, NoPP2, X2I2

Analyses are divided into sections as follows, and only one of these components is specified and included below:

* General characterization of the data, where we visualize various metrics for our cells and/or clusters
* Hierarchical clustering of the data to determine phylogenetic relationship amongst cell clusters
* Cluster-based differential gene expression analysis to determine genes differentially expressed in a cluster relative to the rest of the dataset (cluster-based overall DGE) or in one cluster relative to another (cluster-based pairwise DGE)
* Cluster-based pseudobulk correlation as another method to assess similarity/dissimilarity between pairwise cluster combinations
* Tissue composition comparisons, to determine differences in the compositions of different cells and/or transcripts across different ileal sample types
* Sample-specific quality comparisons, where we assess the whole transcriptomic profiles from each sample
* Cluster-based gene set enrichment analysis, where we calculate enrichment of signatures for sorted immune cell populations from bulk RNA-seq datasets within our single-cell dataset

This file specifically goes over tissue composition comparisons, to determine differences in the compositions of different cells and/or transcripts across different ileal sample types.

## Tissue composition comparisons

### Load required software packages

The following analyses are conducted using base R packages or the scRNA-seq analysis package, Seurat (Stuart and Butler et al. 2019). In addition, we can create density plots by utilizing the software package, ggpointdensity (Kremer 2019). Refer to sessionInfo() at the bottom of the page for the R and package versions used.

Other required packages: ggplot2 (Wickham 2016), reshape2 (Wickham 2007)

```{r, message = FALSE}
library(Seurat)  
library(ggplot2)
library(ggpointdensity) 
library(reshape2)
```

### Import relevant data 

Read in the RDS file from previous work:

```{r, warning = FALSE, error = FALSE}
il <- readRDS("/home/Jayne.Wiarda/scRNAseqIleumAtlas/Ileum/Seurat/IleumAtlasAll.rds") 
DefaultAssay(il) <- "RNA"
Idents(il) <- il$tissue # change identity of each cell to correspond to tissue origin
```

#### Create Seurat objects of only cells derived from each tissue sample type

```{r, warning = FALSE, error = FALSE}
IPPonly <- subset(il, idents = "IPP") # create new object of only IPP tissue cells
NoPPonly <- subset(il, idents = "NoP") # create new object of only NoP tissue cells
I2only <- subset(il, idents = "X2I") # create new object of only X2I tissue cells
Idents(IPPonly) <- IPPonly$seurat_clusters # change identities back to cluster #s
Idents(NoPPonly) <- NoPPonly$seurat_clusters
Idents(I2only) <- I2only$seurat_clusters
```

Visualize cells from each tissue with cell type assignments within UMAP plots:

Cells from IPP samples:

```{r, warning = FALSE, error = FALSE}
DimPlot(IPPonly, 
        reduction = "umap", 
        group.by = "seurat_clusters", 
        label = TRUE)	
```

Cells from NoPP samples:

```{r, warning = FALSE, error = FALSE}
DimPlot(NoPPonly, 
        reduction = "umap", 
        group.by = "seurat_clusters", 
        label = TRUE)	
```

Cells from X2I samples:

```{r, warning = FALSE, error = FALSE}
DimPlot(I2only, 
        reduction = "umap", 
        group.by = "seurat_clusters", 
        label = TRUE)		  	  
```

#### Create density plots to capture cell distributions for each sample type

Create a density plot for cells from IPP samples:

```{r, warning = FALSE, error = FALSE}
IPPembeds = Embeddings(IPPonly[["umap"]]) # extract umap coordinates from Seurat object
IPPembeds <- as.data.frame(IPPembeds) # convert to data frame
ggplot(IPPembeds, aes(x = UMAP_1, y = UMAP_2)) + # create a 2D heatmap
  geom_pointdensity(adjust = 2.47) +
  scale_color_gradientn(colours = c("darkmagenta", "darkturquoise", "yellow", "orange", "red")) + # use custom color gradient scale
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), # set blank white elements
        panel.background = element_blank(), axis.line = element_line(colour = "black"))
```

Create a density plot for cells from NoPP samples:

```{r, warning = FALSE, error = FALSE}
NoPPembeds = Embeddings(NoPPonly[["umap"]]) # extract umap coordinates from Seurat object
NoPPembeds <- as.data.frame(NoPPembeds) # convert to data frame
ggplot(NoPPembeds, aes(x = UMAP_1, y = UMAP_2)) + # create a 2D heatmap
  geom_pointdensity(adjust = 1.65) +
  scale_color_gradientn(colours = c("darkmagenta", "darkturquoise", "yellow", "orange", "red")) + # use custom color gradient scale
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), # set blank white elements
        panel.background = element_blank(), axis.line = element_line(colour = "black"))
```

Create a density plot for cells from X2I samples:

```{r, warning = FALSE, error = FALSE}
I2embeds = Embeddings(I2only[["umap"]]) # extract umap coordinates from seurat object
I2embeds <- as.data.frame(I2embeds) # convert to data frame
ggplot(I2embeds, aes(x = UMAP_1, y = UMAP_2)) + # create a 2D heatmap
  geom_pointdensity(adjust = 2.25) +
  scale_color_gradientn(colours = c("darkmagenta", "darkturquoise", "yellow", "orange", "red")) + # use custom color gradient scale
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), # set blank white elements
        panel.background = element_blank(), axis.line = element_line(colour = "black"))
```

Note that we toggle with the adjust parameter in each plot to make the density scale approximately equal for each of the three plots above.

#### Create pie charts of cell lineage compositions within each tissue type

Swith identities back to the 'celltypes' dataslot that defines our cell lineage assignments:

```{r, warning = FALSE, error = FALSE}
Idents(IPPonly) <- IPPonly$celltypes
Idents(NoPPonly) <- NoPPonly$celltypes
Idents(I2only) <- I2only$celltypes
```

Make a piechart for cells from IPP samples:

```{r, warning = FALSE, error = FALSE}
IPPCellTypeCount <- table(Idents(IPPonly)) # number of cells for each cell type
IPPCellTypePercent <- prop.table(table(Idents(IPPonly))) # frequencies of each cell type
IPPCellType <- as.data.frame(IPPCellTypeCount) #convert to dataframe
IPPCellType <- aggregate(cbind(Freq) ~Var1, data = IPPCellType, FUN = "sum")
cols <- c("mediumorchid", "orange", "blue", "green3")
ggplot(IPPCellType, aes(x=1, y=Freq, fill=Var1))+ # pie chart for IPP cells
  geom_bar(width = 1, stat = "identity") + 
  coord_polar("y", start=0) +
  theme_void() + 
  scale_fill_manual(values=cols)
```

Make a piechart for cells from NoPP samples:

```{r, warning = FALSE, error = FALSE}
NoPPCellTypeCount <- table(Idents(NoPPonly)) # number of cells for each cell type
NoPPCellTypePercent <- prop.table(table(Idents(NoPPonly))) # frequencies of each cell type
NoPPCellType <- as.data.frame(NoPPCellTypeCount) #convert to dataframe
NoPPCellType <- aggregate(cbind(Freq) ~Var1, data = NoPPCellType, FUN = "sum")
#cols <- c("mediumorchid", "orange", "blue", "green3")
ggplot(NoPPCellType, aes(x=1, y=Freq, fill=Var1))+ # pie chart for NoPP cells
  geom_bar(width = 1, stat = "identity") + 
  coord_polar("y", start=0) +
  theme_void() + 
  scale_fill_manual(values=cols)
```

Make a piechart for cells from X2I samples:

```{r, warning = FALSE, error = FALSE}
I2CellTypeCount <- table(Idents(I2only)) # number of cells for each cell type
I2CellTypePercent <- prop.table(table(Idents(I2only))) # frequencies of each cell type
I2CellType <- as.data.frame(I2CellTypeCount) #convert to dataframe
I2CellType <- aggregate(cbind(Freq) ~Var1, data = I2CellType, FUN = "sum")
#cols <- c("mediumorchid", "orange", "blue", "green3")
ggplot(I2CellType, aes(x=1, y=Freq, fill=Var1))+ # pie chart for I2 cells
  geom_bar(width = 1, stat = "identity") + 
  coord_polar("y", start=0) +
  theme_void() + 
  scale_fill_manual(values=cols)
```

### View session information

```{r, warning = FALSE, error = FALSE}
sessionInfo()
```

### References

Lukas P. M. Kremer (2019). ggpointdensity: A Cross Between a 2D Density Plot and a Scatter Plot. R package version 0.1.0. https://CRAN.R-project.org/package=ggpointdensity

Stuart and Butler et al. Comprehensive Integration of Single-Cell Data. Cell (2019).

Hadley Wickham (2007). Reshaping Data with the reshape Package. Journal of Statistical Software, 21(12), 1-20. URL http://www.jstatsoft.org/v21/i12/.