---
title: "Reference-based Label Prediction and Mapping of Porcine Ileal Cells: Creation of a Human Ileum Reference Dataset"
author: "Jayne Wiarda"
date: "5Apr2021"
output: 
  github_document: 
    keep_html: TRUE
---

Since there are already some pretty spectacular and well-annotated scRNA-seq datasets available in literature, we can utilize these external datasets as references we can compare to our porcine ileum scRNA-seq data, which we treat as a dataset to query. We can perform reference-based label transfer to predict the most similar cellular counterparts in already annotated datasets we use as references (Which reference cells are our query cells most similar to?), and we can further calculate mapping scores to see how well query data (our porcine ileum scRNA-seq data) is represented by the reference dataset (How well does the reference data truly resemble the query data?).

Unfortunately, we are the first to create a scRNA-seq dataset of cells in the porcine intestine, so we have to compare to other reference datasets either within the same species but across anatomical locations or within the same tissue but across species. Here are the three relevant datasets we found:

* A scRNA-seq dataset cataloging porcine PBMCs (Herrera-Uribe & Wiarda et al. 2021); referred to as SsPBMC_HerreraWiarda) 
  - Cells are from seven conventional pigs of various ages, including two pigs (ExpD) which are the same as the ones we collected ileum samples from for our ileum scRNA-seq dataset
  - Data can be found at locations specified in the manuscript, but we recovered the data in-house from our previous analyses
  - We removed one cluster labelled as 'Unknown' (cluster 34) prior to creating our reference
* A scRNA-seq dataset cataloging cells from human terminal ileum (Elmentaite et al. 2020; referred to as HsIleum_Elmentaite)
  - Cells are from 8 healthy controls and 7 samples from individuals with Crohn's disease
  - Data were downloaded as pediatric_RAWCOUNTS_cellxgene.h5ad from at https://www.gutcellatlas.org/#fetal
  - Prior to creating the reference, we selected only cells with an inferred state of control for creating our reference, thus excluding cells from Crohn's disease patients, and retained only one-to-one gene orthologs between pig and human
* A scRNA-seq dataset cataloging cells from murine ileum (Xu et al. 2019; referred to as MmIleum_Xu)
  - Cells are from healthy or allergy model mice
  - Data were downloaded from the Single Cell Portal, found at https://singlecell.broadinstitute.org/single_cell/study/SCP210/fasi-immune-mouse-small-intestine#study-summary
  - Prior to creating the reference, we selected only cells from control mice, thus exluding cells from mice modelling an allergic response, and retained only one-to-one gene orthologs between pig and mouse
  
In this section, we go over the steps involved in creating a reference dataset from human ileum.


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.fullwidth=TRUE)
```

### Load required software packages

The R package, Seurat (Stuart & Butler et al. 2019; Hao et al. 2020), is utilized to perform label transfer and mapping. Note, this analysis is done with an upgraded version of Seurat compared to other analyses, and is also performed on another maching. We also utilize the R package, future (Bengtsson 2020), to parallelize some of the steps performed herein. Refer to sessionInfo() at the bottom of the page for the R and package versions used and additional information.

Additional packages used: ggplot2 (Wickham 2016); SeuratDisk (Hoffman 2020)

```{r, message = FALSE}
library(ggplot2)
library(Seurat)
library(future)
library(SeuratDisk)
```

## Create human ileum reference dataset

### Load in gene ortholog information:

Read in human-to-pig gene ortholog information, obtained from biomaRt (Durinck et al. 2009; Durinck et al. 2005; obtained according to instructions outlined at https://www.ensembl.info/2009/01/21/how-to-get-all-the-orthologous-genes-between-two-species/ but specifying human and pig instead of human and mouse), and filter down to only one-to-one gene orthologs:

```{r, warning = FALSE, error = FALSE}
orthoGenes <- read.delim("/Users/jayne.wiarda/Desktop/scRNAseq/scRNAseqIleumAtlas/Ileum/OrthoGenes/PIGvsHUMAN_GeneOrtholog_06022020.txt", # read in gene ortholog file
                         header = TRUE, 
                         skip = 1, 
                         sep = "\t")
orthoGenes <- subset(orthoGenes, Human.homology.type == 'ortholog_one2one') # subset to only one to one orthologs
```

### Load relevant data:

First downloaded pediatric_RAWCOUNTS_cellxgene.h5ad found at https://www.gutcellatlas.org/#fetal; referred to in publication at doi.org/10.1016/j.devcel.2020.11.010. Then, convert the .h5ad file into a Seurat-compatible .h5seurat format:

```{r, warning = FALSE, error = FALSE, message = FALSE}
Convert( # convert downloaded .h5ad formatted dataset into an .h5seurat object. This will be deposited into the same directory as the .h5ad file
  "/Users/jayne.wiarda/Desktop/scRNAseq/scRNAseqIleumAtlas/Ileum/ReferenceData/Elmentaite_HsIleum/pediatric_RAWCOUNTS_cellxgene.h5ad",
  dest = "h5seurat",
  assay = "RNA",
  overwrite = FALSE,
  verbose = TRUE)
HsIleum_Elmentaite <- LoadH5Seurat("/Users/jayne.wiarda/Desktop/scRNAseq/scRNAseqIleumAtlas/Ileum/ReferenceData/Elmentaite_HsIleum/pediatric_RAWCOUNTS_cellxgene.h5seurat") # load in .h5seurat file 
```

### Allow parallelization 

```{r, warning = FALSE, error = FALSE}
plan("multiprocess", workers = 4) # change plan to allow parallelization with 4 workers
options(future.globals.maxSize = 5000 * 1024^2)
```

### Remove cells from Crohn's patients

Only retain cells from healthy controls, which have a 'Control' designation in the inferred.state data slot of the Seurat object:

```{r, warning = FALSE, error = FALSE}
HsIleum_Elmentaite # let's see what we have here...
Idents(HsIleum_Elmentaite) <- HsIleum_Elmentaite$inferred.state # change Seurat identities to inferred disease states already available for cells and used in publication
levels(Idents(HsIleum_Elmentaite)) # list the inferred disease states
HsIleum_Elmentaite <- subset(HsIleum_Elmentaite, 
                             idents = "Control") # subset to take only cells from control samples
HsIleum_Elmentaite # now let's see what we have here after subsetting...we now only have about half as many cells
```

### Filter genes

Filter the genes to include only one-to-one human-to-pig gene orthologs:

```{r, warning = FALSE, error = FALSE}
HsIleum_Elmentaitegenes <- rownames(HsIleum_Elmentaite[['RNA']]@counts) # extract gene names from reference dataset
HsIleum_Elmentaiteorthos <- intersect(HsIleum_Elmentaitegenes, orthoGenes$Human.gene.name) # find which gene names from reference are also one-to-one orthologs
length(HsIleum_Elmentaiteorthos) # how many genes are orthologs?
HsIleum_Elmentaitecounts <- HsIleum_Elmentaite[['RNA']]@counts[rownames(HsIleum_Elmentaite[['RNA']]@counts) %in% HsIleum_Elmentaiteorthos,] # make count matrix from referemce, only taking counts from one-to-one ortholog genes
HsIleum_ElmentaiteMeta <- HsIleum_Elmentaite@meta.data # extract all the meta data from reference
dim(HsIleum_ElmentaiteMeta) # look at dimensions of meta data
```

Create a Seurat object of only our control sample cells with ortholog genes:

```{r, warning = FALSE, error = FALSE}
HsIleum_Elmentaite <- CreateSeuratObject( # now create new Seurat object with only the control cells, orthologous genes, and all meta data
  counts = HsIleum_Elmentaitecounts, 
  meta.data = HsIleum_ElmentaiteMeta)
```

### Re-normalize and re-integrate the data

Split the data into individual samples:

```{r, warning = FALSE, error = FALSE, message = FALSE}
HsIleum_Elmentaite.list <- SplitObject(HsIleum_Elmentaite, split.by = "Sample.name") # split into the original samples that were processed for scRNA-seq
```

For each sample, normalize the data and find the top 2,000 most variable genes:

```{r, warning = FALSE, error = FALSE}
for (i in 1:length(HsIleum_Elmentaite.list)) { # for each sample individually, let's normalize the data and find the 2000 most highly variable features
  HsIleum_Elmentaite.list[[i]] <- NormalizeData(HsIleum_Elmentaite.list[[i]], 
                                                verbose = TRUE, 
                                                normalization.method = "LogNormalize", 
                                                scale.factor = 10000, 
                                                assay = "RNA")
  HsIleum_Elmentaite.list[[i]] <- FindVariableFeatures(HsIleum_Elmentaite.list[[i]], 
                                                       selection.method = "vst", 
                                                       nfeatures = 2000, 
                                                       verbose = TRUE)
}
```

Integrate the samples together:

```{r, warning = FALSE, error = FALSE, message = FALSE}
HsIleum_Elmentaite.anchors <- FindIntegrationAnchors(object.list = HsIleum_Elmentaite.list, 
                                                     dims = 1:30) # find integration anchors between samples based on variable features for each sample with default parameters
referenceHsIleum_Elmentaite <- IntegrateData(anchorset = HsIleum_Elmentaite.anchors, 
                                             dims = 1:30) # integrate the data together based on integration anchors found with default parameters
```

### Re-scale data

```{r, warning = FALSE, error = FALSE, message = FALSE}
referenceHsIleum_Elmentaite <- ScaleData(referenceHsIleum_Elmentaite, 
                                         verbose = TRUE, 
                                         assay = 'integrated') # scale the genes in the integrated assay
```

### Re-run PCA analysis

Calculate the PCs of the integrated dataset:

```{r, warning = FALSE, error = FALSE, message = FALSE}
referenceHsIleum_Elmentaite <- RunPCA(referenceHsIleum_Elmentaite, # calculate first 100 PCs
                                      npcs = 100, 
                                      verbose = TRUE)
```

Visualize our PCs in an elbow plot:

```{r, warning = FALSE, error = FALSE}
ElbowPlot(referenceHsIleum_Elmentaite,
          ndims = 100) # look at this plot to find the 'elbow' for significant PCs... use this number of PCs for creating UMAP, tSNE, & cell neighbors & clustering
```

Quantitiatively calculate how many PCs to use:

```{r, warning = FALSE, error = FALSE}
pct <- referenceHsIleum_Elmentaite[["pca"]]@stdev / sum(referenceHsIleum_Elmentaite[["pca"]]@stdev) * 100 # find standard deviation for each PC
cumu <- cumsum(pct) # find cumulative percentages for PCs
co1 <- which(cumu > 90 & pct < 5)[1] # find PC representing cumulative percent >90% and less than 5% associated with the single PC
co1 # list PC
co2 <- sort(which((pct[1:length(pct) - 1] - pct[2:length(pct)]) > 0.1), decreasing = T)[1] + 1 # find last PC where change in percent variation is more than 0.1%
co2 # list PC
pcs <- min(co1, co2) # find the minimum PC from the 2 methods used above
pcs # list PC
```

Plot our findings:

```{r, warning = FALSE, error = FALSE}
plot_df <- data.frame(pct = pct, # put PC values into dataframe for plotting
                      cumu = cumu, 
                      rank = 1:length(pct))
ggplot(plot_df, aes(cumu, pct, label = rank, color = rank > pcs)) + # visualize PCs to use in elbow plot
  geom_text() + 
  geom_vline(xintercept = 90, color = "grey") + 
  geom_hline(yintercept = min(pct[pct > 5]), color = "grey") +
  theme_bw()
```

Store information for PCs to use in future computations:

```{r, warning = FALSE, error = FALSE}
PCdims <- 1:pcs # use the minimum PC from the quantitative method above to set the PCs for subsequent steps
length(PCdims) # how many significant PCs are there?
```

### Construct UMAP

Create a UMAP, using the calculated number of PCs from above:

```{r, warning = FALSE, error = FALSE, message = FALSE}
referenceHsIleum_Elmentaite <- RunUMAP(referenceHsIleum_Elmentaite, 
                          dims = PCdims, 
                          reduction = "pca", 
                          assay = "SCT") # create UMAP
```

Now check out plotting of some of the metadata groupings:

```{r, warning = FALSE, error = FALSE}
DimPlot(referenceHsIleum_Elmentaite, 
        group.by = "Sample.name") # plot by sample ID
DimPlot(referenceHsIleum_Elmentaite, 
        group.by = "inferred.state") # plot by disease state; should only see control samples
```

```{r, warning = FALSE, error = FALSE, fig.width = 30, fig.height = 15}
DimPlot(referenceHsIleum_Elmentaite, 
        group.by = "annotation_V2", label = TRUE) # plot by cell type annotations assigned in publication
```

### Identify and organize cells 

Change the order clusters are listed in under the annotation_V2 data slot:

```{r, warning = FALSE, error = FALSE}
Idents(referenceHsIleum_Elmentaite) <- referenceHsIleum_Elmentaite$annotation_V2 # change identities to cell annotations from publication
levels(referenceHsIleum_Elmentaite) <- c('enterocyte', 'early enterocyte', 'Goblet cell', 'TA', 'crypt', 'S2 fibroblasts', 'S1 fibroblasts', 'Lymphatic endothelial cell',
                                         'Venous endothelial cell', 'BEST4 enterocyte', 'Paneth cell', 'Glial cell', 'myofibroblast', 'Arterial endothelial cell', 'IL2RG+ enterocyte (M cell)',
                                         'Tuft', 'pericyte', 'S4 fibroblasts', 'enteroendocrine',
                                         'IgA plasma cell', 'IgG plasma cell', 'Memory B cell', 'B cell', 'FCER2 B cell', 'Cycling B cell', 'Activated B cell', 'Cycling plasma cell',
                                         'CD4 T cell', 'CD8 T cell', 'Treg', 'gd T/NK cell', 'Activated T', 'Tfh',
                                         'cDC2', 'Cycling myeloid cells', 'Macrophage', 'cDC1', 'Monocyte', 'pDC', 'activated DC', 'mast cells') # Reorder the clusters based on cell type IDs 
referenceHsIleum_Elmentaite$annotation_V2 <- Idents(referenceHsIleum_Elmentaite) # apply new order to the identities in the Seurat object
```

Now plot these cell classifications on our UMAP:

```{r, warning = FALSE, error = FALSE, fig.width = 30, fig.height = 15}
DimPlot(referenceHsIleum_Elmentaite, # let's see where these cell types fall out in our 2-D visualization plot again
        group.by = "annotation_V2", 
        label = TRUE)
```

Also assign cell lineages to cells:

```{r, warning = FALSE, error = FALSE}
Idents(referenceHsIleum_Elmentaite) <- referenceHsIleum_Elmentaite$annotation_V2
cellLineageList <- c(rep('Non-immune', 19), rep('B lineage', 8), rep('T/ILC lineage', 6), rep('Myeloid lineage', 8))
referenceHsIleum_Elmentaite$cellLineage <- referenceHsIleum_Elmentaite$annotation_V2 # make a new data slot based off cell lineage groupings
Idents(referenceHsIleum_Elmentaite) <- referenceHsIleum_Elmentaite$cellLineage
names(cellLineageList) <- levels(referenceHsIleum_Elmentaite)
referenceHsIleum_Elmentaite <- RenameIdents(referenceHsIleum_Elmentaite, cellLineageList) # change dataset identity to cell lineages
referenceHsIleum_Elmentaite$cellLineage <- Idents(referenceHsIleum_Elmentaite)
```

And plot cell lineage classifications onto the UMAP too:

```{r, warning = FALSE, error = FALSE}
DimPlot(referenceHsIleum_Elmentaite, # let's see where these cell lineages fall out in our 2-D visualization plot again
        group.by = "cellLineage", 
        label = TRUE)
```

### Save the reference dataset as a Seurat object:

```{r, warning = FALSE, error = FALSE}
saveRDS(referenceHsIleum_Elmentaite, '/Users/jayne.wiarda/Desktop/scRNAseq/scRNAseqIleumAtlas/Ileum/ReferenceData/Elmentaite_HsIleum/reference_HsIleum_Elmentaite.rds')
```

### View session information

```{r, warning = FALSE, error = FALSE}
sessionInfo()
```

### References

Mapping identifiers for the integration of genomic datasets with the R/Bioconductor package biomaRt. Steffen Durinck, Paul T. Spellman, Ewan Birney and Wolfgang Huber, Nature Protocols 4, 1184-1191 (2009).
  
BioMart and Bioconductor: a powerful link between biological databases and microarray data analysis. Steffen Durinck, Yves Moreau, Arek Kasprzyk, Sean Davis, Bart De Moor, Alvis Brazma and Wolfgang Huber, Bioinformatics 21, 3439-3440 (2005).

Integrated analysis of multimodal single-cell data. Yuhan Hao, Stephanie Hao, Erica Andersen-Nissen, William M. Mauck III, Shiwei Zheng, Andrew Butler, Maddie J. Lee, Aaron J. Wilk, Charlotte Darby, Michael Zagar, Paul Hoffman, Marlon Stoeckius, Efthymia Papalexi, Eleni P. Mimitou, Jaison Jain, Avi Srivastava, Tim Stuart, Lamar B. Fleming, Bertrand Yeung, Angela J. Rogers, Juliana M. McElrath, Catherine A. Blish, Raphael Gottardo, Peter Smibert, Rahul Satija. bioRxiv 2020.10.12.335331; doi: https://doi.org/10.1101/2020.10.12.335331

Reference transcriptomes of porcine peripheral immune cells created through bulk and single-cell RNA sequencing
Juber Herrera-Uribe, Jayne E Wiarda, Sathesh K Sivasankaran, Lance Daharsh, Haibo Liu, Kristen A Byrne, Timothy Smith, Joan K Lunney, Crystal L Loving, Christopher K Tuggle
bioRxiv 2021.04.02.438107; doi: https://doi.org/10.1101/2021.04.02.438107

Paul Hoffman (2020). SeuratDisk: Interfaces for HDF5-Based Single Cell File Formats. https://mojaveazure.github.io/seurat-disk/, https://github.com/mojaveazure/seurat-disk.

Elmentaite R, Ross ADB, Roberts K, James KR, Ortmann D, Gomes T, Nayak K, Tuck L, Pritchard S, Bayraktar OA, Heuschkel R, Vallier L, Teichmann SA, Zilbauer M. Single-Cell Sequencing of Developing Human Gut Reveals Transcriptional Links to Childhood Crohn's Disease. Dev Cell. 2020 Dec 21;55(6):771-783.e5. doi: 10.1016/j.devcel.2020.11.010.

H. Bengtsson, A Unifying Framework for Parallel and Distributed Processing in R using Futures, arXiv:2008.00553, 2020
 
Stuart and Butler et al. Comprehensive Integration of Single-Cell Data. Cell (2019).

H. Wickham. ggplot2: Elegant Graphics for Data Analysis. Springer-Verlag New York, 2016.
  
Heping Xu, Jiarui Ding, Caroline B.M. Porter, Antonia Wallrapp, Marcin Tabaka, Sai Ma, Shujie Fu, Xuanxuan Guo, Samantha J. Riesenfeld, Chienwen Su, Danielle Dionne, Lan T. Nguyen, Ariel Lefkovith, Orr Ashenberg, Patrick R. Burkett, Hai Ning Shi, Orit Rozenblatt-Rosen, Daniel B. Graham, Vijay K. Kuchroo, Aviv Regev, Ramnik J. Xavier. Transcriptional Atlas of Intestinal Immune Cells Reveals that Neuropeptide α-CGRP Modulates Group 2 Innate Lymphoid Cell Responses. Immunity, Volume 51, Issue 4, 2019, Pages 696-708.e9. doi: 10.1016/j.immuni.2019.09.004.