---
title: "Characterization of Ileal Naive T Cells: Data Subsetting, Normalization, Integration, & Dimensionality Reduction"
author: "Jayne Wiarda"
date: "9Apr2021"
output: 
  github_document: 
    keep_html: TRUE
---

At this point, we've characterized our T/ILC lineage lymphocytes from the porcine ileum scRNA-seq dataset and been able to classify clusters into the following cell types:

* CD4 ab T cells = clusters 12, 15, 26, 41, 46
* Mixed CD4 & CD8 ab T cells (naive) = cluster 24
* CD8 ab T cells = clusters 3, 7, 14, 17, 37
* Mixed gd & CD8 ab T cells = clusters 19, 34, 35, 47
* gd T cells = clusters 6, 23, 29, 31, 32, 51
* ILCs = clusters 1, 18, 43, 44, 53

To define some of these cells further, we want to make the data more manageable to analyze and comprehend, and we will do so by creating subsets of our data based on the following classifications:

* Naive T cells (cluster 24)
* CD4 T cells (clusters 12, 15, 26, 41, 46)
* gd/CD8 ab T cells (clusters 3, 6, 7, 14, 17, 19, 23, 29, 31, 32, 34, 35, 37, 47, 51)
* ILCs (clusters 1, 18, 43, 44, 53)

For this section, we will subset and re-process cells that were identified as naive T cells (cluster 24).

## Data Subsetting, Normalization, Integration, & Dimensionality Reduction

We will first need to normalize our data within each sample, done with the SCTransform method outlined previously (Stuart and Butler et al. 2019). We can then integrate our samples together to correct for confounding technical differences between samples by identifying similar cells across datasets, termed 'anchors' (Stuart and Butler et al. 2019). Next, we perform dimensionality reduction with principle component analysis (PCA) and identify a 'significant' number of principle components (PCs) to use for downstream analyses, such as dimensionality reduction to create uniform manifold approximation and projection (UMAP) and t-distributed stachastic neighbor embedding (t-SNE) visualizations of our data. The method we use to identify 'significant' PCs is outlined previously (https://hbctraining.github.io/scRNA-seq/lessons/elbow_plot_metric.html). We do not re-cluster our data but instead leave original cluster identities intact. The output we create for the dataset is a Seurat object that will be used in many downstream analyses.

We have six ileum samples, with sample nomenclature is as follows:

* IPP = ileum with only Peyer's patch regions present
* NoPP = ileum with only regions without Peyer's patches present
* X2I = ileum with both regions with and without Peyer's patches present; corresponding to a whole ileum cross section
* Pig 1 = IPP1, NoPP1, X2I1
* Pig 2 = IPP2, NoPP2, X2I2

### Load required software packages

The following analyses are conducted primarily using base R packages or the scRNA-seq analysis package, Seurat (Stuart and Butler et al. 2019). To run some of these functions more quickly, we also utilize parallelization with the software package, future (Bengtsson 2020). Refer to sessionInfo() at the bottom of the page for the R and package versions used.

Other required packages: ggplot2 (Wickham 2016), writexl (Ooms 2020), scales (Wickham & Seidel 2020)

```{r, message = FALSE}
library(Seurat)  
library(writexl)        
library(scales)
library(ggplot2)
library(future)
```

### Import relevant data 

Read in the RDS file from previous work:

```{r, warning = FALSE, error = FALSE}
ilT <- readRDS("/home/Jayne.Wiarda/scRNAseqIleumAtlas/Ileum/Seurat/Ileum_TILConly.rds") 
DefaultAssay(ilT) <- "RNA"
```

### Allow parallelization 

```{r, warning = FALSE, error = FALSE}
plan("multiprocess", workers = 4) # change plan to allow parallelization with 4 workers
options(future.globals.maxSize = 5000 * 1024^2)
```

### Subset data

Create a subset of data containing only naive T cells:

```{r, warning = FALSE, error = FALSE}
Idents(ilT) <- ilT$TILCtypes # change cell identities to the T/ILC types we specified above, stored under the 'TILCtypes' metadata slot of our Seurat object
ilTnaive <- subset(ilT, idents = 'CD4/CD8 mix') # create a new dataset of only cells identified as T/ILC lineage lymphocytes
rm(ilT) # remove the Seurat object for the entire dataset... we no longer need this
```

### Remove unnecessary data

Identify genes with non-zero expression to keep:

```{r, warning = FALSE, error = FALSE}
counts <- as.data.frame(ilTnaive[['RNA']]@counts)
keep <- rowSums(counts) > 0
keep <- rownames(counts[keep,])
```

Slim down our seurat object:

```{r, warning = FALSE, error = FALSE}
ilTnaive <- DietSeurat(ilTnaive, 
                  counts = TRUE,
                  data = TRUE,
                  scale.data = FALSE, # remove the scaled data
                  dimreducs = NULL,
                  features = keep, # keep only genes with non-zero counts across all cells
                  assays = 'RNA') # keep only RNA assay and remove SCT and integrated
```

### Re-normalize the data

Re-perform SCTransform on each individual sample:

```{r, warning = FALSE, error = FALSE, message = FALSE}
ilTnaive.list <- SplitObject(ilTnaive, split.by = "SampleID") # split by sample IDs
ilTnaive.list <- ilTnaive.list[c("IPP1", "IPP2", "NoPP1", "NoPP2", "X2I1", "X2I2")] 
for (i in 1:length(ilTnaive.list)) { # normalize data using SCTransform method
  ilTnaive.list[[i]] <- SCTransform(ilTnaive.list[[i]], 
                               return.only.var.genes = FALSE, 
                               verbose = TRUE) 
}
```

### Re-integrate the data

Re-integrate samples based on re-calculated SCT assays:

```{r, warning = FALSE, error = FALSE, message = FALSE}
ilTnaive.features <- SelectIntegrationFeatures(ilTnaive.list, # select the genes to use for integration
                                          verbose = TRUE) 
ilTnaive.list <- PrepSCTIntegration(ilTnaive.list, 
                               anchor.features = ilTnaive.features,
                               verbose = TRUE)
ilTnaive.anchors <- FindIntegrationAnchors(ilTnaive.list, # identify anchors for integration from top 30 PCs
                                      normalization.method = "SCT", 
                                      anchor.features = ilTnaive.features, 
                                      dims = 1:30,
                                      k.filter = 40) # since we have so few cells in the dataset, we have to lower our k.filter number. We tried different k.filter values in multiples of 5, and 40 is the highest number for which we could still find integration anchors.
ilTnaive.integrated <- IntegrateData(ilTnaive.anchors, # integrate data
                                normalization.method = "SCT", 
                                dims = 1:30)
```

### Re-run PCA analysis

Calculate the PCs of the integrated dataset:

```{r, warning = FALSE, error = FALSE, messge = FALSE}
ilTnaive.integrated <- RunPCA(ilTnaive.integrated, # run PCA analysis for 100 dimensions of the data
                         npcs = 100, 
                         verbose = TRUE) 
```

Visualize our PCs in an elbow plot:

```{r, warning = FALSE, error = FALSE}
ElbowPlot(ilTnaive.integrated,
          ndims = 100) # look at this plot to find the 'elbow' for significant PCs... use this number of PCs for creating UMAP, tSNE, & cell neighbors & clustering
```

Quantitiatively calculate how many PCs to use:

```{r, warning = FALSE, error = FALSE}
pct <- ilTnaive.integrated[["pca"]]@stdev / sum(ilTnaive.integrated[["pca"]]@stdev) * 100 # find standard deviation for each PC
cumu <- cumsum(pct) # find cumulative percentages for PCs
co1 <- which(cumu > 90 & pct < 5)[1] # find PC representing cumulative percent >90% and less than 5% associated with the single PC
co1 # list PC
co2 <- sort(which((pct[1:length(pct) - 1] - pct[2:length(pct)]) > 0.1), decreasing = T)[1] + 1 # find last PC where change in percent variation is more than 0.1%
co2 # list PC
pcs <- min(co1, co2) # find the minimum PC from the 2 methods used above
pcs # list PC
```

Plot our findings:

```{r, warning = FALSE, error = FALSE}
plot_df <- data.frame(pct = pct, # put PC values into dataframe for plotting
                      cumu = cumu, 
                      rank = 1:length(pct))
ggplot(plot_df, aes(cumu, pct, label = rank, color = rank > pcs)) + # visualize PCs to use in elbow plot
  geom_text() + 
  geom_vline(xintercept = 90, color = "grey") + 
  geom_hline(yintercept = min(pct[pct > 5]), color = "grey") +
  theme_bw()
```

Store information for PCs to use in future computations:

```{r, warning = FALSE, error = FALSE}
PCdims <- 1:pcs 
PCdims
```

### Construct UMAP and t-SNE plots

```{r, warning = FALSE, error = FALSE, message = FALSE}
ilTnaive.integrated <- RunUMAP(ilTnaive.integrated, 
                          dims = PCdims, 
                          reduction = "pca", 
                          assay = "SCT") # create UMAP
ilTnaive.integrated <- RunTSNE(ilTnaive.integrated, 
                          dims = PCdims, 
                          reduction = "pca", 
                          assay = "SCT") # create tSNE plot (if desired)
```

Plot UMAP:

```{r, warning = FALSE, error = FALSE}
DimPlot(ilTnaive.integrated, 
        label = TRUE,
        reduction = 'umap',
        group.by = "seurat_clusters") # view UMAP plot
```

Plot t-SNE:

```{r, warning = FALSE, error = FALSE}
DimPlot(ilTnaive.integrated, 
        label = TRUE,
        reduction = 'tsne',
        group.by = "seurat_clusters") # view t-SNE plot
```

### Add normalized and scaled data to the RNA assay

At this point, our Seurat object has three 'assays', where we store counts information. The SCT assay has counts normalized and scaled with SCTransform, the integration assay has counts data used to integrate data, and the RNA assay, which at this point, still needs to be normalized and scaled.

Since it's still recommended to use RNA counts rather than SCT counts for many downstream analyses, we need to normalize and scale our raw RNA counts.

Normalize the RNA data with standard log normalization:

```{r, warning = FALSE, error = FALSE}
ilTnaive.integrated <- NormalizeData(ilTnaive.integrated,  
                                normalization.method = "LogNormalize", 
                                scale.factor = 10000, 
                                assay = "RNA")
```

Scale the RNA data: 

```{r, warning = FALSE, error = FALSE, message = FALSE}
ilTnaive.integrated <- ScaleData(ilTnaive.integrated, 
                            assay = "RNA")
```

See that we now have scaled RNA data:

```{r, warning = FALSE, error = FALSE}
dim(ilTnaive.integrated[["RNA"]]@scale.data) 
```

### Save the Seurat object:

Now, importantly, we save our Seurat object as an .rds. We will end up importing the .rds back into R for many downstream analyses.

```{r, warning = FALSE, error = FALSE}
DefaultAssay(ilTnaive.integrated) <- 'RNA'
saveRDS(ilTnaive.integrated, file = "/home/Jayne.Wiarda/scRNAseqIleumAtlas/Ileum/Seurat/Ileum_TnaiveOnly.rds") # save .rds file
```

### View session information

```{r, warning = FALSE, error = FALSE}
sessionInfo()
```

### References

H. Bengtsson, A Unifying Framework for Parallel and Distributed Processing in R using Futures, arXiv:2008.00553, 2020

Jeroen Ooms (2020). writexl: Export Data Frames to Excel 'xlsx' Format. R package version 1.3.1. https://CRAN.R-project.org/package=writexl
  
Stuart and Butler et al. Comprehensive Integration of Single-Cell Data. Cell (2019).

H. Wickham. ggplot2: Elegant Graphics for Data Analysis. Springer-Verlag New York, 2016.
  
Hadley Wickham and Dana Seidel (2020). scales: Scale Functions for Visualization. R package version 1.1.1. https://CRAN.R-project.org/package=scales