---
title: "Annotation of gd/CD8 T cells"
author: "Jayne Wiarda"
date: "27May2021"
output: 
  github_document: 
    keep_html: TRUE
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.fullwidth=TRUE)
```

We have now processed non-naive gd/CD8 ab T cells from porcine ileum using three methods: cell clustering (and accompanying DGE and hierarchical clustering analyses), topic modeling, and multidimensional DGE analysis. Based on these results, we create final annotations for our cells in this script.

### Load required software packages

The following analyses are conducted using base R packages or the scRNA-seq analysis package, Seurat (Stuart and Butler et al. 2019), ggplot2 (Wickham 2016), scales (Wickham & Seidel 2020), dplyr (Wickham et al. 2020), readxl (Wickham & Bryan 2019). Refer to sessionInfo() at the bottom of the page for the R and package versions used.

```{r, message = FALSE}
library(Seurat)
library(ggplot2)
library(scales)
library(dplyr)
library(readxl)
```

### Import relevant data

Read in our Seurat object from previous analysis:

```{r, warning = FALSE, error = FALSE}
il <- readRDS("/home/Jayne.Wiarda/scRNAseqIleumAtlas/Ileum/Seurat/Ileum_gdCD8Tonly.rds") # read in corresponding Seurat object
Idents(il) <- il$seurat_clusters
levels(il) <- c('3', '6', '7', '14', '17', '19', '23', '29', '31', '32', '34', '35', '37', '47', '51') 
il$neworder <- Idents(il)
```

## Visualize our cell clusters and their relationships

Start by re-visualizing clusters on a t-SNE plot:

```{r, warning = FALSE, error = FALSE}
DimPlot(il, 
        reduction = 'tsne', 
        group.by = 'neworder',
        label = TRUE)
```

Then re-construct cluster heirarchy:

```{r, warning = FALSE, error = FALSE}
data.tree <- Tool(object = il, 
                  slot = "BuildClusterTree") 
data.tree <- ape::rotateConstr(data.tree, c('51', '32', '35', '23', '37', '3', '14', 
                                            '19', '47', '17', '7', '34', '6', '29', '31'))
plot(data.tree, direction = 'downwards', edge.width = 1.5, font = 1)
```

And also re-visualize DGE across clusters:

```{r, warning = FALSE, error = FALSE, message = FALSE}
DE <- read.table('/home/Jayne.Wiarda/scRNAseqIleumAtlas/Ileum/DE/gdCD8TOnly_OverallDE.txt')
DE <- subset(DE, avg_logFC > 0) # only take genes enriched in the clusters 
Idents(il) <- il$phyloorder
topgenes <- DE %>% group_by(cluster) %>% top_n(5, avg_logFC) # only plot top 5 genes per cluster, as determined by highest average logFC values
DoHeatmap(subset(il, downsample = 100), # take only 100 cells per cluster for plotting
          features = as.character(topgenes$gene), 
          assay = "RNA", 
          disp.min = -1.5, 
          disp.max = 3) +
  scale_fill_gradientn(colors = c("darkturquoise", "grey90", "indianred1", "red"))
```
        
## Visualize our topic model weights & multidimensional DGE gene module scores

Let's read and organize the data:

```{r, warning = FALSE, error = FALSE}
fits <- readRDS('/home/Jayne.Wiarda/scRNAseqIleumAtlas/Ileum/fastTopics/Ileum_gdCD8Tonly_K_3_K_10_topic_model_fits.rds') # read in all the topic models
cellMem <- as.data.frame(fits$k3$L) # find weighted topic membership for cells when using 3 topics
spatialMod <- as.data.frame(read_excel("/home/Jayne.Wiarda/scRNAseqIleumAtlas/Ileum/Haystack/GDCD8Tonly_k4_DetectionScores.xlsx", col_types = c(rep('numeric',4), 'text')))
rownames(spatialMod) <- spatialMod$CellBarcodes
spatialMod$CellBarcodes <- NULL
il <- AddMetaData(il, metadata = c(cellMem, spatialMod)) # add topic membership and gene module enrichment to Seurat object
noK <- paste0('k', 1:3)
noMod <- paste0('GeneModule', 1:4)
```

Overlay topic weights onto t-SNE plot:

```{r, warning = FALSE, error = FALSE, message = FALSE}
FeaturePlot(il, features = c(noK), 
            reduction = 'tsne',
            ncol = 3) & 
  scale_color_gradientn( colours = c('grey90', 'darkgreen'),  limits = c(0, 1), oob = squish) & 
  NoAxes() & NoLegend()
```

Overlay gene module detection scores onto t-SNE plot:

```{r, warning = FALSE, error = FALSE, message = FALSE}
FeaturePlot(il, features = c(noMod), 
            reduction = 'tsne',
            ncol = 3) & 
  scale_color_gradientn( colours = c('grey90', 'red4'),  limits = c(0, 0.6), oob = squish) & 
  NoAxes() & NoLegend()
```

## Isolate highly unique clusters of SELLhi gd T cells and CD2- gd T cells

From looking at cluster-based DGE & hierarchy, we can recognize two highly unique clusters that aren't well-captured by gene modules or topics. These are cluster 51, a population of gd T cells with high expression of SELL, and cluster 31, a population of CD2- gd T cells. We will isolate these two populations from the rest of the data and call them SELLhi gd T and CD2neg gd T, respectively.

Show where SELLhi gd T cells are on t-SNE:

```{r, warning = FALSE, error = FALSE}
SELLhiCells <- rownames(il@meta.data %>% filter(il$seurat_clusters == '51'))
DimPlot(il, 
        cells.highlight = SELLhiCells, 
        cols.highlight = "red", 
        sizes.highlight = 1,
        cols = "black", 
        order = TRUE, 
        reduction = 'tsne') & NoAxes() & NoLegend() # SELLhi gd T cells in red
```

Show where CD2- gd T cells are on t-SNE:

```{r, warning = FALSE, error = FALSE}
CD2negCells <- rownames(il@meta.data %>% filter(il$seurat_clusters == '31'))
DimPlot(il, 
        cells.highlight = CD2negCells, 
        cols.highlight = "red", 
        sizes.highlight = 1,
        cols = "black", 
        order = TRUE, 
        reduction = 'tsne') & NoAxes() & NoLegend() # CD2neg gd T cells in red
```

## Identify cycling gd/CD8 T cells

Start by removing our identified CD2neg and SELLhi gd T cells from the dataset, so we are now only working with the remaining un-annotated cells:

```{r, warning = FALSE, error = FALSE}
Idents(il) <- il$neworder
il2 <- subset(il, idents = c('31', '51'), invert = TRUE)
```

From looking at genes contributing to topics and gene modules and accompanying GO results, we can see that gene module 3 and topic 3 are enriched for genes required for cellular replication/division, and thus gene module 3 and topic 3 are higher in cycling cells. Based on this information, let's gate out cycling cells that had high topic 3 weights and/or high gene module 2 detection scores.

Highlight cycling cells in red as those with gene module 3 detection scores > 0.11 and/or topic 3 weight > 0.41:

```{r, warning = FALSE, error = FALSE}
plot(il2$k3, il2$GeneModule3, pch = 20, col= ifelse(il2$k3 > 0.41 | il2$GeneModule3 > 0.11, 'red', 'black')) # cycling cells in red
```

Show where cycling cells are found in our t-SNE plot:

```{r, warning = FALSE, error = FALSE}
RepCells <- rownames(il2@meta.data[il2$k3 > 0.41 | il2$GeneModule3 > 0.11 , ])
DimPlot(il, 
        cells.highlight = RepCells, 
        cols.highlight = "red", 
        sizes.highlight = 1,
        cols = "black", 
        order = TRUE, 
        reduction = 'tsne') & NoAxes() & NoLegend() # cycling cells in red
```

## Identify cytotoxic T cells

Remove annotated cells from the dataset, so we are now only working with the remaining un-annotated cells:

```{r, warning = FALSE, error = FALSE}
il3 <- il2[,!colnames(il2) %in% RepCells]
```

From looking at genes contributing to topics and gene modules and accompanying GO results, we can see that gene module 1 and topic 1 are enriched for genes typically expressed by activated T cells. Moreover, topic 2 had an inverse pattern of enrichment and was enriched for genes/processes characteristic of cell cytotoxicity. Based on this information, let's gate out activated but not largely cytotoxic T cells that had higher topic 1 than topic 2 weights and/or high gene module 1 detection scores. Remaining cells can then be classified as cytotoxic T cells.

Start by visualizing topic 1 weights vs gene module 1 detection scores, overlaid with topic 2 weights:

```{r, warning = FALSE, error = FALSE}
ggplot(il3@meta.data, aes(x=k1, y=GeneModule1, color = k2)) + 
  geom_point() + 
  scale_colour_gradient(low='grey90', high='darkgreen') +
  theme_classic()
```

Highlight activated cells in red as those with topic 1 at least 4x higher than topic 2 weights and/or gene module 1 detection scores > 0.25:

```{r, warning = FALSE, error = FALSE}
plot(il3$k1, il3$GeneModule1, pch = 20, col= ifelse(il3$k1 > 4*il3$k2 | il3$GeneModule1 > 0.25 , "red", "black"))
```

Show where cytotoxic cells are found in our t-SNE plot:

```{r, warning = FALSE, error = FALSE}
ActCells <- rownames(il3@meta.data %>% filter(il3$k1 > 4*il3$k2 | il3$GeneModule1 > 0.25))
DimPlot(il3, 
        cells.highlight = ActCells, 
        cols.highlight = "red", 
        sizes.highlight = 1,
        cols = "black", 
        order = TRUE, 
        reduction = 'tsne') # activated cells in red
```

## Identify activated T cells

We identify the remaining T cells as cytotoxic T cells due to high weighting of topic 2 and expression of many genes assocaited with cellular cytotoxicity.

Show where remaining cytotoxic T cells are found in our t-SNE plot:

```{r, warning = FALSE, error = FALSE}
CytotoxCells <- rownames(il3@meta.data %>% filter(!(il3$k1 > 4*il3$k2 | il3$GeneModule1 > 0.25)))
DimPlot(il3, 
        cells.highlight = CytotoxCells, 
        cols.highlight = "red", 
        sizes.highlight = 1,
        cols = "black", 
        order = TRUE, 
        reduction = 'tsne') & NoAxes() & NoLegend() # cytotoxic cells in red
```

## Identify gd vs CD8 T cells

We have now obtained three classifications that contain a mixture of both gd and CD8 ab T cells, including cycling, cytotoxic, and activated cells. Let's now divide these three cell types further into gd vs CD8 ab T cells by assessing expression of gd marker TRDC and CD8 ab T cell marker CD8B.

View potential co-expression of TRDC and CD8B... we see markers are largely mutually exclusive, as expected:

```{r, warning = FALSE, error = FALSE}
FeaturePlot(il,
            features = c('CD8B', 'TRDC'),
            cols = c("beige", "red3", "dodgerblue3"),
            reduction = 'tsne',
            blend = TRUE,
            blend.threshold = 0.1, 
            min.cutoff = c(0,0),
            max.cutoff = c(3.5, 5.5))
```

Calculate the ratio of TRDC/CD8B and CD8B/TRDC:

```{r, warning = FALSE, error = FALSE}
TRDCvCD8B <- as.data.frame(il[['RNA']]@data["TRDC",] + 1) / (il[['RNA']]@data["CD8B",] + 1)
CD8BvTRDC <- as.data.frame(il[['RNA']]@data["CD8B",] + 1) / (il[['RNA']]@data["TRDC",] + 1)
```

Based on the ratios, we will classify gd T cells as those cells that had more TRDC transcripts than CD8B transcripts. Cells that either had more CD8B than TRDC transcripts or had sum zero of CD8B or TRDC will be classified as CD8 ab T cells. Cells with sum zero transcripts were still considered CD8 rather than gd T cells because TRDC expression was observed to be less sparse in our dataset than CD8B, making it more likely a cell not having TRDC transcript is a true negative.

Plot the ratios, highlighting gd T cells in red:

```{r, warning = FALSE, error = FALSE}
plot(TRDCvCD8B[,1], CD8BvTRDC[,1], pch = 20, col= ifelse(TRDCvCD8B[,1] > 1 & CD8BvTRDC[,1] <= 1, "red", "black")) # this is how to form 1:1 ratio cut-off
```

Add the ratios to our metadata:

```{r, warning = FALSE, error = FALSE}
il <- AddMetaData(il, metadata = c(TRDCvCD8B, CD8BvTRDC)) # add TRDC vs CD8B cell metadata in Seurat object
```

Show where gd T cells are found in our t-SNE plot:

```{r, warning = FALSE, error = FALSE}
gdCells <- rownames(il@meta.data %>% filter(il$il...RNA....data..TRDC.......1 > 1 & il$il...RNA....data..CD8B.......1 <= 1))
DimPlot(il, 
        cells.highlight = gdCells, 
        cols.highlight = "red", 
        sizes.highlight = 1,
        cols = "black", 
        order = TRUE, 
        reduction = 'tsne') & NoAxes() & NoLegend() # gd cells in red
```

Show where CD8 ab T cells are found in our t-SNE plot:

```{r, warning = FALSE, error = FALSE}
CD8Cells <- rownames(il@meta.data %>% filter(!(il$il...RNA....data..TRDC.......1 > 1 & il$il...RNA....data..CD8B.......1 <= 1)))
DimPlot(il, 
        cells.highlight = CD8Cells, 
        cols.highlight = "red", 
        sizes.highlight = 1,
        cols = "black", 
        order = TRUE, 
        reduction = 'tsne') & NoAxes() & NoLegend() # CD8 ab T cells in red
```

Now split cycling, cytotoxic, and activated cells into gd vs CD8 populations:

```{r, warning = FALSE, error = FALSE}
RepGD <- intersect(gdCells, RepCells)
ActGD <- intersect(gdCells, ActCells)
toxGD <- intersect(gdCells, CytotoxCells)
RepCD8 <- intersect(CD8Cells, RepCells)
ActCD8 <- intersect(CD8Cells, ActCells)
toxCD8 <- intersect(CD8Cells, CytotoxCells)
```

## Re-visualize data with new annotations

Now let's add the new annotations to our meta data in the Seurat object:

```{r, warning = FALSE, error = FALSE}
bcs <- as.data.frame(colnames(il))
colnames(bcs) <- 'barcode'
bcs$cellID <- rep('_', nrow(bcs)) # if we left any cells un-annotated, they will show up with an ID of '_'
bcs <- bcs %>% mutate(cellID = replace(cellID, barcode %in% CD2negCells, 'CD2neg gd T'))
bcs <- bcs %>% mutate(cellID = replace(cellID, barcode %in% SELLhiCells, 'SELLhi gd T'))
bcs <- bcs %>% mutate(cellID = replace(cellID, barcode %in% RepGD, 'Cycling gd T'))
bcs <- bcs %>% mutate(cellID = replace(cellID, barcode %in% ActGD, 'Activated gd T'))
bcs <- bcs %>% mutate(cellID = replace(cellID, barcode %in% toxGD, 'Cytotoxic gd T'))
bcs <- bcs %>% mutate(cellID = replace(cellID, barcode %in% RepCD8, 'Cycling CD8 ab T'))
bcs <- bcs %>% mutate(cellID = replace(cellID, barcode %in% ActCD8, 'Activated CD8 ab T'))
bcs <- bcs %>% mutate(cellID = replace(cellID, barcode %in% toxCD8, 'Cytotoxic CD8 ab T'))
write.table(bcs, '/home/Jayne.Wiarda/scRNAseqIleumAtlas/Ileum/Annotations/gdCD8T.txt')
rownames(bcs) <- bcs$barcode
bcs$barcode <- NULL
il <- AddMetaData(il, metadata = bcs) # add topic membership and gene module enrichment to Seurat object
```

Plot the annotated populations:

```{r, warning = FALSE, error = FALSE}
DimPlot(il, 
        group.by = 'cellID', 
        reduction = 'tsne', 
        cols = c('yellow', 'darkorange', 'sandybrown', 'yellow4', 'chocolate4', 'goldenrod1', 'darkorange3', 'tan'))
```

And let's also highlight only one cell populations at a time...

Activated CD8 ab T cells:

```{r, warning = FALSE, error = FALSE}
DimPlot(il, 
        cells.highlight = ActCD8, 
        cols.highlight = "yellow", 
        sizes.highlight = 1,
        cols = "grey80", 
        order = TRUE, 
        reduction = 'tsne') & NoAxes() & NoLegend()
```

Activated gd T cells:

```{r, warning = FALSE, error = FALSE}
DimPlot(il, 
        cells.highlight = ActGD, 
        cols.highlight = "darkorange", 
        sizes.highlight = 1,
        cols = "grey80", 
        order = TRUE, 
        reduction = 'tsne') & NoAxes() & NoLegend()
```

CD2neg gd T cells:

```{r, warning = FALSE, error = FALSE}
DimPlot(il, 
        cells.highlight = CD2negCells, 
        cols.highlight = "sandybrown", 
        sizes.highlight = 1,
        cols = "grey80", 
        order = TRUE, 
        reduction = 'tsne') & NoAxes() & NoLegend()
```

Cycling CD8 ab T cells:

```{r, warning = FALSE, error = FALSE}
DimPlot(il, 
        cells.highlight = RepCD8, 
        cols.highlight = "yellow4", 
        sizes.highlight = 1,
        cols = "grey80", 
        order = TRUE, 
        reduction = 'tsne') & NoAxes() & NoLegend()
```

Cycling gd cells:

```{r, warning = FALSE, error = FALSE}
DimPlot(il, 
        cells.highlight = RepGD, 
        cols.highlight = "chocolate4", 
        sizes.highlight = 1,
        cols = "grey80", 
        order = TRUE, 
        reduction = 'tsne') & NoAxes() & NoLegend()
```

Cytotoxic CD8 ab cells:

```{r, warning = FALSE, error = FALSE}
DimPlot(il, 
        cells.highlight = toxCD8, 
        cols.highlight = "goldenrod1", 
        sizes.highlight = 1,
        cols = "grey80", 
        order = TRUE, 
        reduction = 'tsne') & NoAxes() & NoLegend()
```

Cytotoxic gd cells:

```{r, warning = FALSE, error = FALSE}
DimPlot(il, 
        cells.highlight = toxGD, 
        cols.highlight = "darkorange3", 
        sizes.highlight = 1,
        cols = "grey80", 
        order = TRUE, 
        reduction = 'tsne') & NoAxes() & NoLegend()
```

SELLhi gd cells:

```{r, warning = FALSE, error = FALSE}
DimPlot(il, 
        cells.highlight = SELLhiCells, 
        cols.highlight = "tan", 
        sizes.highlight = 1,
        cols = "grey80", 
        order = TRUE, 
        reduction = 'tsne') & NoAxes() & NoLegend()
```

### Save data

Lastly, let's re-save our Seurat object with the new annotations stored in the cellID meta data slot:

```{r, warning = FALSE, error = FALSE, message = FALSE}
Idents(il) <- il$cellID
saveRDS(il, '/home/Jayne.Wiarda/scRNAseqIleumAtlas/Ileum/Seurat/Ileum_gdCD8Tonly.rds')
```

### View session information

```{r, warning = FALSE, error = FALSE}
sessionInfo()
```

### References
  
Stuart and Butler et al. Comprehensive Integration of Single-Cell Data. Cell (2019).

H. Wickham. ggplot2: Elegant Graphics for Data Analysis. Springer-Verlag New York, 2016.

Hadley Wickham and Jennifer Bryan (2019). readxl: Read Excel Files. R package version 1.3.1.
  https://CRAN.R-project.org/package=readxl
  
Hadley Wickham and Dana Seidel (2020). scales: Scale Functions for Visualization. R package version 1.1.1.
  https://CRAN.R-project.org/package=scales
  
Hadley Wickham, Romain François, Lionel Henry and Kirill Müller (2020). dplyr: A Grammar of Data Manipulation.
  R package version 1.0.2. https://CRAN.R-project.org/package=dplyr
