---
title: "Annotation of ILCs"
author: "Jayne Wiarda"
date: "27May2021"
output: 
  github_document: 
    keep_html: TRUE
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.fullwidth=TRUE)
```

We have now processed ILCs from porcine ileum using three methods: cell clustering (and accompanying DGE and hierarchical clustering analyses), topic modeling, and multidimensional DGE analysis. Based on these results, we create final annotations for our cells in this script.

### Load required software packages

The following analyses are conducted using base R packages or the scRNA-seq analysis package, Seurat (Stuart and Butler et al. 2019), ggplot2 (Wickham 2016), scales (Wickham & Seidel 2020), dplyr (Wickham et al. 2020), readxl (Wickham & Bryan 2019). Refer to sessionInfo() at the bottom of the page for the R and package versions used.

```{r, message = FALSE}
library(Seurat)
library(ggplot2)
library(scales)
library(dplyr)
library(readxl)
```

### Import relevant data

Read in our Seurat object from previous analysis:

```{r, warning = FALSE, error = FALSE}
il <- readRDS('/home/Jayne.Wiarda/scRNAseqIleumAtlas/Ileum/Seurat/Ileum_ILConly.rds')
Idents(il) <- il$seurat_clusters
```

## Visualize our cell clusters and their relationships

Start by re-visualizing clusters on a t-SNE plot:

```{r, warning = FALSE, error = FALSE}
DimPlot(il, 
        reduction = 'tsne', 
        group.by = 'neworder',
        label = TRUE)
```

Then re-construct cluster heirarchy:

```{r, warning = FALSE, error = FALSE}
data.tree <- Tool(object = il, 
                  slot = "BuildClusterTree") 
data.tree <- ape::rotateConstr(data.tree, c('43', '44', '1', '18', '53'))
plot(data.tree, direction = 'downwards', edge.width = 1.5, font = 1)
```

And also re-visualize DGE across clusters:

```{r, warning = FALSE, error = FALSE, message = FALSE}
DE <- read.table('/home/Jayne.Wiarda/scRNAseqIleumAtlas/Ileum/DE/ILConly_OverallDE.txt')
DE <- subset(DE, avg_logFC > 0) # only take genes enriched in the clusters 
Idents(il) <- il$phyloorder
topgenes <- DE %>% group_by(cluster) %>% top_n(5, avg_logFC) # only plot top 5 genes per cluster, as determined by highest average logFC values
DoHeatmap(subset(il, downsample = 100), # take only 100 cells per cluster for plotting
          features = as.character(topgenes$gene), 
          assay = "RNA", 
          disp.min = -1.5, 
          disp.max = 3) +
  scale_fill_gradientn(colors = c("darkturquoise", "grey90", "indianred1", "red"))
```
        
## Visualize our topic model weights & multidimensional DGE gene module scores

Let's read and organize the data:

```{r, warning = FALSE, error = FALSE}
fits <- readRDS('/home/Jayne.Wiarda/scRNAseqIleumAtlas/Ileum/fastTopics/Ileum_ILConly_K_3_K_10_topic_model_fits.rds') # read in all the topic models
cellMem <- as.data.frame(fits$k3$L) # find weighted topic membership for cells when using 3 topics
spatialMod <- as.data.frame(read_excel("/home/Jayne.Wiarda/scRNAseqIleumAtlas/Ileum/Haystack/ILConly_k3_DetectionScores.xlsx", col_types = c(rep('numeric',3), 'text')))
rownames(spatialMod) <- spatialMod$CellBarcodes
spatialMod$CellBarcodes <- NULL
il <- AddMetaData(il, metadata = c(cellMem, spatialMod)) # add topic membership and gene module enrichment to Seurat object
noK <- paste0('k', 1:3)
noMod <- paste0('GeneModule', 1:3)
```

Overlay topic weights onto t-SNE plot:

```{r, warning = FALSE, error = FALSE, message = FALSE}
FeaturePlot(il, features = c(noK), 
            reduction = 'tsne',
            ncol = 3) & 
  scale_color_gradientn( colours = c('grey90', 'darkgreen'),  limits = c(0, 1), oob = squish) & 
  NoAxes() & NoLegend()
```

Overlay gene module detection scores onto t-SNE plot:

```{r, warning = FALSE, error = FALSE, message = FALSE}
FeaturePlot(il, features = c(noMod), 
            reduction = 'tsne',
            ncol = 3) & 
  scale_color_gradientn( colours = c('grey90', 'red4'),  limits = c(0, 0.8), oob = squish) & 
  NoAxes() & NoLegend()
```

## Identify group 3 ILCs

From looking at cluster-based DGE & hierarchy, we recognize cluster 43 to be most distantly related by cluster-based hierarchy. We also see cluster 43 has generally high topic 1 weights and gene module 1 detection scores, while having low gene module 3 detection. We see topic 1 and gene module 1 are associated with many genes of type 3 immunity, while gene module 3 is associated with many genes of type 1 immunity. Plot topic 1, gene module 1, and gene module 3:

Plot topic 1, gene module 1, and gene module 3:

```{r, warning = FALSE, error = FALSE}
ggplot(il@meta.data, aes(x=k1, y=GeneModule3, color = GeneModule1)) + 
  geom_point() + 
  scale_colour_gradient(low='grey90', high='red4') +
  theme_classic()
```

Now let's see where cluster 43 cells fall out on the above plot:

```{r, warning = FALSE, error = FALSE}
plot(il$k1, il$GeneModule3, pch = 20, col= ifelse(il$seurat_clusters == '43', "red", "black")) # group 3 ILCs in red
```

Based on the plot, we will classify cells in cluster 43 as group 1 ILCs.

Show where group 3 ILCs are on t-SNE:

```{r, warning = FALSE, error = FALSE}
ILC3 <- rownames(il@meta.data %>% filter(il$seurat_clusters == '43'))
DimPlot(il, 
        cells.highlight = ILC3, 
        cols.highlight = "red", 
        sizes.highlight = 1,
        cols = "black", 
        order = TRUE, 
        reduction = 'tsne') & NoAxes() & NoLegend() # ILC3s in red
```

## Identify cycling group 1 ILCs

Start by removing our identified group 3 ILCs from the dataset, so we are now only working with the remaining un-annotated cells, which we are calling group 1 ILCs due to high gene module 3 detection scores:

```{r, warning = FALSE, error = FALSE}
Idents(il) <- il$seurat_clusters
il2 <- subset(il, idents = c('43'), invert = TRUE)
```

From looking at cluster-based DGE, we see cluster 53 has a distinct profile characteristic of cycling cells. Therefore, we will classify cells in cluster 53 as cycling group 1 ILCs.

Show where cycling group 1 ILCs are on t-SNE:

```{r, warning = FALSE, error = FALSE}
repILC1 <- rownames(il@meta.data %>% filter(il$seurat_clusters == '53'))
DimPlot(il, 
        cells.highlight = repILC1, 
        cols.highlight = "red", 
        sizes.highlight = 1,
        cols = "black", 
        order = TRUE, 
        reduction = 'tsne') & NoAxes() & NoLegend() # cycling group 1 ILCs in red
```

## Identify activated group 1 ILCs

Remove annotated cells from the dataset, so we are now only working with the remaining un-annotated cells:

```{r, warning = FALSE, error = FALSE}
Idents(il2) <- il2$seurat_clusters
il3 <- subset(il2, idents = c('53'), invert = TRUE)
```

From looking at genes contributing to topics and gene modules and accompanying GO results, we can see that gene module 2 and topic 2 are enriched for genes typically expressed by activated cells. In addition, topic 3 is enriched for many genes associated with cellular cytotoxicty and tends to have inverse expression patterns to topic 2/gene module 2. Based on this information, let's identify activated group 1 ILCs that have high topic 2 weights/gene module 2 detection scores. Remaining cells can be called cytotoxic group 1 ILCs due to high topic 3 weights.

Start by visualizing topic 3 weights vs gene module 2 detection scores, overlaid with topic 2 weights:

```{r, warning = FALSE, error = FALSE}
ggplot(il3@meta.data, aes(x=k3, y=GeneModule2, color = k2)) + 
  geom_point() + 
  scale_colour_gradient(low='grey90', high='darkgreen') +
  theme_classic()
```

Highlight activated group 1 ILCs in red as those with topic 3 weights < 0.05 and gene module 2 detection scores > 0.4, or topic 2 weights > 0.9:

```{r, warning = FALSE, error = FALSE}
plot(il3$k3, il3$GeneModule2, pch = 20, col= ifelse(il3$k3 < 0.05 & il3$GeneModule2 > 0.4 | il3$k2 > 0.9, "red", "black")) # activated group 1 ILCs in red
```

Show where activated cells are found in our t-SNE plot:

```{r, warning = FALSE, error = FALSE}
actILC1 <- rownames(il3@meta.data %>% filter(il3$k3 < 0.05 & il3$GeneModule2 > 0.4 | il3$k2 > 0.9))
DimPlot(il3, 
        cells.highlight = actILC1, 
        cols.highlight = "red", 
        sizes.highlight = 1,
        cols = "black", 
        order = TRUE, 
        reduction = 'tsne') # activated group 1 ILCs in red
```

## Identify cytotoxic group 1 ILCs

We identify the remaining cells as cytotoxic group 1 ILCs due to high weighting of topic 3, associated with many genes indicitive of cytotoxicity.

Show where remaining cytotoxic cells are found in our t-SNE plot:

```{r, warning = FALSE, error = FALSE}
cytotoxILC1 <- rownames(il3@meta.data %>% filter(!(il3$k3 < 0.05 & il3$GeneModule2 > 0.4 | il3$k2 > 0.9)))
DimPlot(il3, 
        cells.highlight = cytotoxILC1, 
        cols.highlight = "red", 
        sizes.highlight = 1,
        cols = "black", 
        order = TRUE, 
        reduction = 'tsne') & NoAxes() & NoLegend() # cytotoxic group 1 ILCs in red
```

## Re-visualize data with new annotations

Now let's add the new annotations to our meta data in the Seurat object:

```{r, warning = FALSE, error = FALSE}
bcs <- as.data.frame(colnames(il))
colnames(bcs) <- 'barcode'
bcs$cellID <- rep('_', nrow(bcs)) # if we left any cells un-annotated, they will show up with an ID of '_'
bcs <- bcs %>% mutate(cellID = replace(cellID, barcode %in% ILC3, 'Group 3 ILCs'))
bcs <- bcs %>% mutate(cellID = replace(cellID, barcode %in% repILC1, 'Cycling group 1 ILCs'))
bcs <- bcs %>% mutate(cellID = replace(cellID, barcode %in% actILC1, 'Activated group 1 ILCs'))
bcs <- bcs %>% mutate(cellID = replace(cellID, barcode %in% cytotoxILC1, 'Cytotoxic group 1 ILCs'))
write.table(bcs, '/home/Jayne.Wiarda/scRNAseqIleumAtlas/Ileum/Annotations/ILC.txt')
rownames(bcs) <- bcs$barcode
bcs$barcode <- NULL
il <- AddMetaData(il, metadata = bcs) # add topic membership and gene module enrichment to Seurat object
```

Plot the annotated populations:

```{r, warning = FALSE, error = FALSE}
DimPlot(il, 
        group.by = 'cellID', 
        reduction = 'tsne', 
        cols = c('red', 'darkred', 'indianred1', 'black'))
```

And let's also highlight only one cell populations at a time...

Activated group 1 ILCs:

```{r, warning = FALSE, error = FALSE}
DimPlot(il, 
        cells.highlight = actILC1, 
        cols.highlight = "red", 
        sizes.highlight = 1,
        cols = "grey80", 
        order = TRUE, 
        reduction = 'tsne') & NoAxes() & NoLegend()
```

Cytotoxic group 1 ILCs:

```{r, warning = FALSE, error = FALSE}
DimPlot(il, 
        cells.highlight = cytotoxILC1, 
        cols.highlight = "indianred1", 
        sizes.highlight = 1,
        cols = "grey80", 
        order = TRUE, 
        reduction = 'tsne') & NoAxes() & NoLegend()
```

Cycling group 1 ILCs:

```{r, warning = FALSE, error = FALSE}
DimPlot(il, 
        cells.highlight = repILC1, 
        cols.highlight = "darkred", 
        sizes.highlight = 1,
        cols = "grey80", 
        order = TRUE, 
        reduction = 'tsne') & NoAxes() & NoLegend()
```

Group 3 ILCs:

```{r, warning = FALSE, error = FALSE}
DimPlot(il, 
        cells.highlight = ILC3, 
        cols.highlight = "black", 
        sizes.highlight = 1,
        cols = "grey80", 
        order = TRUE, 
        reduction = 'tsne') & NoAxes() & NoLegend()
```

### Save data

Lastly, let's re-save our Seurat object with the new annotations stored in the cellID meta data slot:

```{r, warning = FALSE, error = FALSE, message = FALSE}
Idents(il) <- il$cellID
saveRDS(il, '/home/Jayne.Wiarda/scRNAseqIleumAtlas/Ileum/Seurat/Ileum_ILConly.rds')
```

### View session information

```{r, warning = FALSE, error = FALSE}
sessionInfo()
```

### References
  
Stuart and Butler et al. Comprehensive Integration of Single-Cell Data. Cell (2019).

H. Wickham. ggplot2: Elegant Graphics for Data Analysis. Springer-Verlag New York, 2016.

Hadley Wickham and Jennifer Bryan (2019). readxl: Read Excel Files. R package version 1.3.1.
  https://CRAN.R-project.org/package=readxl
  
Hadley Wickham and Dana Seidel (2020). scales: Scale Functions for Visualization. R package version 1.1.1.
  https://CRAN.R-project.org/package=scales
  
Hadley Wickham, Romain François, Lionel Henry and Kirill Müller (2020). dplyr: A Grammar of Data Manipulation.
  R package version 1.0.2. https://CRAN.R-project.org/package=dplyr
